// @ariakit/react-components/dist/hovercard/hovercard-disclosure.js
"use client";
import { useDialogDisclosure } from "../dialog/dialog-disclosure.js";
import { useVisuallyHidden } from "../visually-hidden/visually-hidden.js";
import { useHovercardProviderContext } from "./hovercard-context.js";
import { createElement, createHook, forwardRef, useEvent, useMergeRefs } from "@ariakit/react-utils";
import { addGlobalEventListener, contains, invariant, isElement, isNode } from "@ariakit/utils";
import { useEffect, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { sync } from "@ariakit/store";
//#region src/hovercard/hovercard-disclosure.tsx
const TagName = "button";
/**
* Returns props to create a `HovercardDisclosure` component.
* @see https://ariakit.com/components/hovercard
* @example
* ```jsx
* const store = useHovercardStore();
* const props = useHovercardDisclosure({ store });
* <HovercardAnchor store={store}>@username</HovercardAnchor>
* <Role {...props} />
* <Hovercard store={store}>Details</Hovercard>
* ```
*/
const useHovercardDisclosure = createHook(function useHovercardDisclosure({ store, ...props }) {
	const context = useHovercardProviderContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "HovercardDisclosure must receive a `store` prop or be wrapped in a HovercardProvider component.");
	const [visible, setVisible] = useState(false);
	useEffect(() => {
		if (!visible) return;
		const onBlur = (event) => {
			if (!store) return;
			const nextActiveElement = event.relatedTarget;
			if (isNode(nextActiveElement)) {
				const { anchorElement: anchor, popoverElement: popover, disclosureElement: disclosure } = store.getState();
				if (anchor && contains(anchor, nextActiveElement)) return;
				if (popover && contains(popover, nextActiveElement)) return;
				if (disclosure && contains(disclosure, nextActiveElement)) return;
				if (isElement(nextActiveElement) && nextActiveElement.hasAttribute("data-focus-trap")) return;
			}
			setVisible(false);
		};
		return addGlobalEventListener("focusout", onBlur, true);
	}, [visible, store]);
	useEffect(() => {
		return sync(store, ["anchorElement"], (state) => {
			const anchor = state.anchorElement;
			if (!anchor) return;
			const observer = new MutationObserver(() => {
				if (!anchor.hasAttribute("data-focus-visible")) return;
				setVisible(true);
			});
			observer.observe(anchor, { attributeFilter: ["data-focus-visible"] });
			return () => observer.disconnect();
		});
	}, [store]);
	const onClickProp = props.onClick;
	const onClick = useEvent((event) => {
		onClickProp?.(event);
		if (event.defaultPrevented) return;
		store?.setAutoFocusOnShow(true);
	});
	const onFocusProp = props.onFocus;
	const onFocus = useEvent((event) => {
		onFocusProp?.(event);
		if (event.defaultPrevented) return;
		setVisible(true);
	});
	const { style } = useVisuallyHidden();
	if (!visible) props = {
		...props,
		style: {
			...style,
			...props.style
		}
	};
	props = {
		children: /* @__PURE__ */ jsx("svg", {
			display: "block",
			fill: "none",
			stroke: "currentColor",
			strokeLinecap: "round",
			strokeLinejoin: "round",
			strokeWidth: 1.5,
			viewBox: "0 0 16 16",
			height: "1em",
			width: "1em",
			children: /* @__PURE__ */ jsx("polyline", { points: "4,6 8,10 12,6" })
		}),
		...props,
		ref: useMergeRefs(store.setDisclosureElement, props.ref),
		onClick,
		onFocus
	};
	props = useDialogDisclosure({
		store,
		...props
	});
	return props;
});
/**
* Renders a hidden disclosure button that will be visible when the
* [`HovercardAnchor`](https://ariakit.com/reference/hovercard-anchor) receives
* keyboard focus. The user can then navigate to the button to open the
* hovercard when using the keyboard.
* @see https://ariakit.com/components/hovercard
* @example
* ```jsx {3}
* <HovercardProvider>
*   <HovercardAnchor>@username</HovercardAnchor>
*   <HovercardDisclosure />
*   <Hovercard>Details</Hovercard>
* </HovercardProvider>
* ```
*/
const HovercardDisclosure = forwardRef(function HovercardDisclosure(props) {
	return createElement(TagName, useHovercardDisclosure(props));
});
//#endregion
export { HovercardDisclosure, useHovercardDisclosure };
