// @ariakit/react-components/dist/menubar/menubar.js
"use client";
import { useComposite } from "../composite/composite.js";
import { MenubarScopedContextProvider, useMenubarProviderContext } from "./menubar-context.js";
import { useMenubarStore } from "./menubar-store.js";
import { createElement, createHook, forwardRef, useWrapElement } from "@ariakit/react-utils";
import { jsx } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
//#region src/menubar/menubar.tsx
const TagName = "div";
/**
* Returns props to create a `Menubar` component.
* @see https://ariakit.com/components/menubar
* @example
* ```jsx
* const store = useMenubarStore();
* const menubarProps = useMenubar({ store });
* const fileProps = useMenuItem({ store });
* const fileMenu = useMenuStore();
* <Role {...menubarProps}>
*   <MenuButton {...fileProps} store={fileMenu}>
*     File
*   </MenuButton>
*   <Menu store={fileMenu}>
*     <MenuItem>New File</MenuItem>
*     <MenuItem>New Window</MenuItem>
*   </Menu>
* </Role>
* ```
*/
const useMenubar = createHook(function useMenubar({ store: storeProp, composite = true, orientation: orientationProp, virtualFocus, focusLoop, rtl, ...props }) {
	const context = useMenubarProviderContext();
	storeProp = storeProp || context;
	const store = useMenubarStore({
		store: storeProp,
		orientation: orientationProp,
		virtualFocus,
		focusLoop,
		rtl
	});
	const orientation = useStoreState(store, (state) => !composite || state.orientation === "both" ? void 0 : state.orientation);
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(MenubarScopedContextProvider, {
		value: store,
		children: element
	}), [store]);
	if (composite) props = {
		role: "menubar",
		"aria-orientation": orientation,
		...props
	};
	props = useComposite({
		store,
		composite,
		...props
	});
	return props;
});
/**
* Renders a menubar that may contain a group of
* [`MenuItem`](https://ariakit.com/reference/menu-item) elements that control
* other submenus.
* @see https://ariakit.com/components/menubar
* @example
* ```jsx
* <Menubar>
*   <MenuProvider>
*     <MenuItem render={<MenuButton />}>File</MenuItem>
*     <Menu>
*       <MenuItem>New File</MenuItem>
*       <MenuItem>New Window</MenuItem>
*     </Menu>
*   </MenuProvider>
*   <MenuProvider>
*     <MenuItem render={<MenuButton />}>Edit</MenuItem>
*     <Menu>
*       <MenuItem>Undo</MenuItem>
*       <MenuItem>Redo</MenuItem>
*     </Menu>
*   </MenuProvider>
* </Menubar>
* ```
*/
const Menubar = forwardRef(function Menubar(props) {
	return createElement(TagName, useMenubar(props));
});
//#endregion
export { Menubar, useMenubar };
