// codemirror-graphql/utils/mode-factory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_language_service_1 = require("graphql-language-service");
const mode_indent_1 = __importDefault(require("./mode-indent"));
const graphqlModeFactory = config => {
    const parser = (0, graphql_language_service_1.onlineParser)({
        eatWhitespace: stream => stream.eatWhile(graphql_language_service_1.isIgnored),
        lexRules: graphql_language_service_1.LexRules,
        parseRules: graphql_language_service_1.ParseRules,
        editorConfig: { tabSize: config.tabSize },
    });
    return {
        config,
        startState: parser.startState,
        token: parser.token,
        indent: mode_indent_1.default,
        electricInput: /^\s*[})\]]/,
        fold: 'brace',
        lineComment: '#',
        closeBrackets: {
            pairs: '()[]{}""',
            explode: '()[]{}',
        },
    };
};
exports.default = graphqlModeFactory;
