// @ariakit/react-components/dist/tag/utils.js
"use client";
import { isTouchDevice, toArray } from "@ariakit/utils";
import { useEffect, useState } from "react";
//#region src/tag/utils.ts
const DEFAULT_DELIMITER = [
	"\n",
	";",
	",",
	/\s/
];
function getDelimiters(delimiter, defaultDelimiter = DEFAULT_DELIMITER) {
	const finalDelimiter = delimiter === void 0 ? defaultDelimiter : delimiter;
	if (!finalDelimiter) return [];
	return toArray(finalDelimiter);
}
function splitValueByDelimiter(value, delimiters) {
	for (const delimiter of delimiters) {
		let match = matchDelimiter(value, delimiter);
		while (match?.index === 0 && match.length > 0) {
			value = value.slice(match.length);
			match = matchDelimiter(value, delimiter);
		}
		if (!match) continue;
		return value.split(delimiter);
	}
	return [];
}
function matchDelimiter(value, delimiter) {
	if (typeof delimiter === "string") {
		const index = value.indexOf(delimiter);
		if (index === -1) return null;
		return {
			index,
			length: delimiter.length
		};
	}
	const match = value.match(delimiter);
	if (!match) return null;
	return {
		index: match.index,
		length: match[0].length
	};
}
function useTouchDevice() {
	const [touchDevice, setTouchDevice] = useState(false);
	useEffect(() => {
		setTouchDevice(isTouchDevice());
	}, []);
	return touchDevice;
}
//#endregion
export { getDelimiters, splitValueByDelimiter, useTouchDevice };
