// @ariakit/react-components/dist/toolbar/toolbar.js
"use client";
import { useComposite } from "../composite/composite.js";
import { ToolbarScopedContextProvider, useToolbarProviderContext } from "./toolbar-context.js";
import { useToolbarStore } from "./toolbar-store.js";
import { createElement, createHook, forwardRef, useWrapElement } from "@ariakit/react-utils";
import { jsx } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
//#region src/toolbar/toolbar.tsx
const TagName = "div";
/**
* Returns props to create a `Toolbar` component.
* @see https://ariakit.com/components/toolbar
* @example
* ```jsx
* const store = useToolbarStore();
* const props = useToolbar({ store });
* <Role {...props}>
*   <ToolbarItem>Item 1</ToolbarItem>
*   <ToolbarItem>Item 2</ToolbarItem>
* </Role>
* ```
*/
const useToolbar = createHook(function useToolbar({ store: storeProp, orientation: orientationProp, virtualFocus, focusLoop, rtl, ...props }) {
	const context = useToolbarProviderContext();
	storeProp = storeProp || context;
	const store = useToolbarStore({
		store: storeProp,
		orientation: orientationProp,
		virtualFocus,
		focusLoop,
		rtl
	});
	const orientation = useStoreState(store, (state) => state.orientation === "both" ? void 0 : state.orientation);
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(ToolbarScopedContextProvider, {
		value: store,
		children: element
	}), [store]);
	props = {
		role: "toolbar",
		"aria-orientation": orientation,
		...props
	};
	props = useComposite({
		store,
		...props
	});
	return props;
});
/**
* Renders a toolbar element that groups interactive elements together.
* @see https://ariakit.com/components/toolbar
* @example
* ```jsx
* <Toolbar>
*   <ToolbarItem>Item 1</ToolbarItem>
*   <ToolbarItem>Item 2</ToolbarItem>
* </Toolbar>
* ```
*/
const Toolbar = forwardRef(function Toolbar(props) {
	return createElement(TagName, useToolbar(props));
});
//#endregion
export { Toolbar, useToolbar };
