// @ariakit/react-components/dist/collection/collection-renderer.js
"use client";
import { useCollectionContext } from "./collection-context.js";
import { createElement, forwardRef, useBooleanEvent, useEvent, useForceUpdate, useId, useMergeRefs, useWrapElement } from "@ariakit/react-utils";
import { getScrollingElement, getWindow, invariant, shallowEqual } from "@ariakit/utils";
import { createContext, useCallback, useContext, useEffect, useMemo, useRef, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
import { flushSync } from "react-dom";
//#region src/collection/collection-renderer.tsx
const TagName = "div";
const CollectionRendererContext = createContext(null);
function createTask() {
	let raf = 0;
	const run = (cb) => {
		if (raf) return;
		raf = requestAnimationFrame(() => {
			raf = 0;
			cb();
		});
	};
	const cancel = () => {
		cancelAnimationFrame(raf);
		raf = 0;
	};
	return {
		run,
		cancel
	};
}
function findNearestIndex(items, target, getValue) {
	let left = 0;
	let right = getItemsLength(items) - 1;
	while (left <= right) {
		const index = (left + right) / 2 | 0;
		const value = getValue(index);
		if (value === target) return index;
		else if (value < target) left = index + 1;
		else right = index - 1;
	}
	if (left > 0) return left - 1;
	return 0;
}
function getItemsLength(items) {
	return typeof items === "number" ? items : items.length;
}
function getItemObject(item) {
	if (!item || typeof item !== "object") return { value: item };
	return item;
}
function getItemId(item, index, baseId) {
	invariant(baseId, "CollectionRenderer must be given an `id` prop.");
	if (item && typeof item === "object") {
		const itemObject = getItemObject(item);
		if (itemObject.id != null) return itemObject.id;
	}
	return `${baseId}/${index}`;
}
function getItem(items, index) {
	if (typeof items === "number") {
		if (index >= items) return null;
		return {};
	}
	const item = items[index];
	if (!item) return null;
	if (typeof item === "object") return item;
	return { value: item };
}
function getItemSize(item, horizontal, fallbackElement) {
	const itemObject = getItemObject(item);
	const prop = horizontal ? "width" : "height";
	const style = itemObject.style;
	if (style) {
		const size = style[prop];
		if (typeof size === "number") return size;
	}
	const items = itemObject.items;
	const hasSameOrientation = !itemObject.orientation || horizontal && itemObject.orientation === "horizontal" || !horizontal && itemObject.orientation === "vertical";
	if (items?.length && hasSameOrientation) {
		const padding = (itemObject.paddingStart ?? itemObject.padding ?? 0) + (itemObject.paddingEnd ?? itemObject.padding ?? 0);
		const initialSize = (itemObject.gap ?? 0) * (items.length - 1) + padding;
		if (itemObject.itemSize) return initialSize + itemObject.itemSize * items.length;
		const totalSize = items.reduce((sum, item) => sum + getItemSize(item, horizontal), initialSize);
		if (totalSize !== initialSize) return totalSize;
	}
	const element = fallbackElement !== false ? itemObject.element || fallbackElement : null;
	if (element?.isConnected) return element.getBoundingClientRect()[prop];
	if (items?.length && !hasSameOrientation) {
		const maxSize = items.reduce((max, item) => Math.max(max, getItemSize(item, horizontal)), 0);
		if (maxSize) return maxSize;
	}
	return 0;
}
function getAverageSize(props) {
	const length = getItemsLength(props.items);
	let currentIndex = 0;
	let averageSize = props.estimatedItemSize;
	const setAverageSize = (size) => {
		const prevIndex = currentIndex;
		currentIndex = currentIndex + 1;
		averageSize = (averageSize * prevIndex + size) / currentIndex;
	};
	for (let index = 0; index < length; index += 1) {
		const item = getItem(props.items, index);
		const itemId = getItemId(item, index, props.baseId);
		const itemData = props.data.get(itemId);
		const fallbackElement = props.elements.get(itemId);
		const size = getItemSize(item, props.horizontal, fallbackElement);
		if (size) setAverageSize(size);
		else if (itemData?.rendered) setAverageSize(itemData.end - itemData.start);
	}
	return averageSize;
}
function getScrollOffset(scroller, horizontal) {
	if ("scrollX" in scroller) return horizontal ? scroller.scrollX : scroller.scrollY;
	return horizontal ? scroller.scrollLeft : scroller.scrollTop;
}
function getViewport(scroller) {
	const { defaultView, documentElement } = scroller.ownerDocument;
	if (scroller === documentElement) return defaultView;
	return scroller;
}
function useScroller(rendererRef) {
	const [scroller, setScroller] = useState(null);
	useEffect(() => {
		const renderer = rendererRef?.current;
		if (!renderer) return;
		const scroller = getScrollingElement(renderer);
		if (!scroller) return;
		setScroller(scroller);
	}, [rendererRef]);
	return scroller;
}
function getRendererOffset(renderer, scroller, horizontal) {
	const win = getWindow(renderer);
	const htmlElement = win?.document.documentElement;
	const rendererRect = renderer.getBoundingClientRect();
	const rendererOffset = horizontal ? rendererRect.left : rendererRect.top;
	if (scroller === htmlElement) return getScrollOffset(win, horizontal) + rendererOffset;
	const scrollerRect = scroller.getBoundingClientRect();
	const scrollerOffset = horizontal ? scrollerRect.left : scrollerRect.top;
	const scrollOffset = getScrollOffset(scroller, horizontal);
	return rendererOffset - scrollerOffset + scrollOffset;
}
function getOffsets(renderer, scroller, horizontal) {
	const scrollOffset = getScrollOffset(scroller, horizontal);
	const rendererOffset = getRendererOffset(renderer, scroller, horizontal);
	const scrollSize = horizontal ? scroller.clientWidth : scroller.clientHeight;
	const start = scrollOffset - rendererOffset;
	return {
		start,
		end: start + scrollSize
	};
}
function getItemsEnd(props) {
	const length = getItemsLength(props.items);
	const totalPadding = props.paddingStart + props.paddingEnd;
	if (!length) return totalPadding;
	const lastIndex = length - 1;
	const totalGap = lastIndex * props.gap;
	if (props.itemSize != null) return length * props.itemSize + totalGap + totalPadding;
	const defaultEnd = length * props.estimatedItemSize + totalGap + totalPadding;
	if (!props.baseId) return defaultEnd;
	const lastItemId = getItemId(getItem(props.items, lastIndex), lastIndex, props.baseId);
	const lastItemData = props.data.get(lastItemId);
	if (lastItemData?.end) return lastItemData.end + props.paddingEnd;
	if (!Array.isArray(props.items)) return defaultEnd;
	const end = props.items.reduce((sum, item) => sum + getItemSize(item, props.horizontal, false), 0);
	if (!end) return defaultEnd;
	return end + totalGap + totalPadding;
}
function getData(props) {
	const length = getItemsLength(props.items);
	let nextData;
	let start = props.paddingStart;
	const avgSize = getAverageSize(props);
	for (let index = 0; index < length; index += 1) {
		const item = getItem(props.items, index);
		const itemId = getItemId(item, index, props.baseId);
		const itemData = props.data.get(itemId);
		const prevRendered = itemData?.rendered ?? false;
		const setSize = (size, rendered = prevRendered) => {
			if (index > 0) start += props.gap;
			const end = start + size;
			if (!shallowEqual(itemData, {
				index,
				rendered,
				start,
				end
			})) {
				if (!nextData) nextData = new Map(props.data);
				nextData.set(itemId, {
					index,
					rendered,
					start,
					end
				});
			}
			start = end;
		};
		const size = getItemSize(item, props.horizontal, props.elements.get(itemId));
		if (size) setSize(size, true);
		else if (itemData?.rendered) setSize(itemData.end - itemData.start, true);
		else setSize(avgSize);
	}
	return nextData;
}
function useCollectionRenderer({ store, items: itemsProp, initialItems = 0, gap = 0, itemSize, estimatedItemSize = 40, overscan: overscanProp, orientation: orientationProp, padding = 0, paddingStart = padding, paddingEnd = padding, persistentIndices, renderOnScroll = true, renderOnResize = !!renderOnScroll, children: renderItem, ...props }) {
	const context = useCollectionContext();
	store = store || context;
	const items = useStoreState(store, (state) => itemsProp ?? state?.items);
	invariant(items != null, process.env.NODE_ENV !== "production" && "CollectionRenderer must be either wrapped in a Collection component or be given an `items` prop.");
	let parent = useContext(CollectionRendererContext);
	if (store && parent?.store !== store) parent = null;
	const parentData = parent?.childrenData;
	const orientation = orientationProp ?? parent?.orientation ?? "vertical";
	const overscan = overscanProp ?? parent?.overscan ?? 1;
	const ref = useRef(null);
	const baseId = useId(props.id);
	const horizontal = orientation === "horizontal";
	const elements = useMemo(() => /* @__PURE__ */ new Map(), []);
	const [elementsUpdated, updateElements] = useForceUpdate();
	const computeData = useCallback((currentData, currentBaseId, currentItems) => {
		return getData({
			baseId: currentBaseId,
			items: currentItems,
			data: currentData,
			gap,
			elements,
			horizontal,
			paddingStart,
			estimatedItemSize
		});
	}, [
		gap,
		elements,
		horizontal,
		paddingStart,
		estimatedItemSize
	]);
	const [defaultVisibleIndices, setVisibleIndices] = useState(() => {
		if (!initialItems) return [];
		const length = getItemsLength(items);
		const initialLength = Math.min(length, Math.abs(initialItems));
		return Array.from({ length: initialLength }, (_, index) => {
			if (initialItems < 0) return length - index - 1;
			return index;
		});
	});
	const visibleIndices = useMemo(() => {
		if (!persistentIndices) return defaultVisibleIndices;
		const nextIndices = defaultVisibleIndices.slice();
		for (const index of persistentIndices) {
			if (index < 0) continue;
			if (nextIndices.includes(index)) continue;
			nextIndices.push(index);
		}
		nextIndices.sort((a, b) => a - b);
		if (shallowEqual(defaultVisibleIndices, nextIndices)) return defaultVisibleIndices;
		return nextIndices;
	}, [defaultVisibleIndices, persistentIndices]);
	const [data, setData] = useState(() => {
		if (!baseId) return /* @__PURE__ */ new Map();
		const data = parentData?.get(baseId) || /* @__PURE__ */ new Map();
		if (itemSize != null) return data;
		if (!items) return data;
		return computeData(data, baseId, items) || data;
	});
	const totalSize = useMemo(() => {
		return getItemsEnd({
			baseId,
			items,
			data,
			gap,
			horizontal,
			itemSize,
			estimatedItemSize,
			paddingStart,
			paddingEnd
		});
	}, [
		baseId,
		items,
		data,
		gap,
		horizontal,
		itemSize,
		estimatedItemSize,
		paddingStart,
		paddingEnd
	]);
	useEffect(() => {
		if (!baseId) return;
		parentData?.set(baseId, data);
	}, [
		baseId,
		parentData,
		data
	]);
	useEffect(() => {
		if (itemSize != null) return;
		if (!baseId) return;
		if (!items) return;
		const nextData = computeData(data, baseId, items);
		if (nextData) setData(nextData);
	}, [
		elementsUpdated,
		itemSize,
		baseId,
		items,
		data,
		computeData
	]);
	const scroller = useScroller(items ? ref : null);
	const offsetsRef = useRef({
		start: 0,
		end: 0
	});
	const processVisibleIndices = useCallback(() => {
		const offsets = offsetsRef.current;
		if (!items) return;
		if (!baseId) return;
		if (!offsets.end) return;
		if (!data.size && !itemSize) return;
		const length = getItemsLength(items);
		const getItemOffset = (index, prop = "start") => {
			if (itemSize) {
				const start = itemSize * index + gap * index + paddingStart;
				if (prop === "start") return start;
				return start + itemSize;
			}
			const itemId = getItemId(getItem(items, index), index, baseId);
			return data.get(itemId)?.[prop] ?? 0;
		};
		const initialStart = findNearestIndex(items, offsets.start, getItemOffset);
		let initialEnd = initialStart;
		while (initialEnd < length && getItemOffset(initialEnd) < offsets.end) initialEnd += 1;
		const finalOverscan = initialEnd - initialStart ? overscan : 0;
		const start = Math.max(initialStart - finalOverscan, 0);
		const end = Math.min(initialEnd + finalOverscan, length);
		const indices = Array.from({ length: end - start }, (_, index) => index + start);
		setVisibleIndices((prevIndices) => {
			if (shallowEqual(prevIndices, indices)) return prevIndices;
			return indices;
		});
	}, [
		elementsUpdated,
		items,
		baseId,
		data,
		itemSize,
		gap,
		paddingStart,
		overscan
	]);
	useEffect(processVisibleIndices, [processVisibleIndices]);
	const processVisibleIndicesEvent = useEvent(processVisibleIndices);
	useEffect(() => {
		const renderer = ref.current;
		if (!renderer) return;
		if (!scroller) return;
		offsetsRef.current = getOffsets(renderer, scroller, horizontal);
		processVisibleIndicesEvent();
	}, [
		scroller,
		horizontal,
		processVisibleIndicesEvent
	]);
	const mayRenderOnScroll = !!renderOnScroll;
	const renderOnScrollProp = useBooleanEvent(renderOnScroll);
	useEffect(() => {
		if (!mayRenderOnScroll) return;
		const renderer = ref.current;
		if (!renderer) return;
		if (!scroller) return;
		const viewport = getViewport(scroller);
		if (!viewport) return;
		const task = createTask();
		const onScroll = (event) => {
			task.run(() => {
				if (!renderOnScrollProp(event)) return;
				offsetsRef.current = getOffsets(renderer, scroller, horizontal);
				processVisibleIndicesEvent();
			});
		};
		viewport.addEventListener("scroll", onScroll, { passive: true });
		return () => {
			task.cancel();
			viewport.removeEventListener("scroll", onScroll);
		};
	}, [
		mayRenderOnScroll,
		scroller,
		renderOnScrollProp,
		horizontal,
		processVisibleIndicesEvent
	]);
	const mayRenderOnResize = !!renderOnResize;
	const renderOnResizeProp = useBooleanEvent(renderOnResize);
	useEffect(() => {
		if (!mayRenderOnResize) return;
		const renderer = ref.current;
		if (!renderer) return;
		if (!scroller) return;
		const viewport = getViewport(scroller);
		if (!viewport) return;
		const task = createTask();
		if (viewport === scroller) {
			if (typeof ResizeObserver !== "function") return;
			let firstRun = true;
			const observer = new ResizeObserver(() => {
				if (firstRun) {
					firstRun = false;
					return;
				}
				task.run(() => {
					if (!renderOnResizeProp(scroller)) return;
					offsetsRef.current = getOffsets(renderer, scroller, horizontal);
					processVisibleIndicesEvent();
				});
			});
			observer.observe(scroller);
			return () => {
				task.cancel();
				observer.disconnect();
			};
		}
		const onResize = () => {
			task.run(() => {
				if (!renderOnResizeProp(scroller)) return;
				offsetsRef.current = getOffsets(renderer, scroller, horizontal);
				processVisibleIndicesEvent();
			});
		};
		viewport.addEventListener("resize", onResize, { passive: true });
		return () => {
			task.cancel();
			viewport.removeEventListener("resize", onResize);
		};
	}, [
		mayRenderOnResize,
		scroller,
		renderOnResizeProp,
		horizontal,
		processVisibleIndicesEvent
	]);
	useEffect(() => {
		if (typeof IntersectionObserver !== "function") return;
		const renderer = ref.current;
		if (!renderer) return;
		if (!scroller) return;
		const viewport = getViewport(scroller);
		if (!viewport) return;
		const observer = new IntersectionObserver(() => {
			offsetsRef.current = getOffsets(renderer, scroller, horizontal);
			processVisibleIndicesEvent();
		}, { root: scroller === viewport ? scroller : null });
		observer.observe(renderer);
		return () => {
			observer.disconnect();
		};
	}, [
		scroller,
		horizontal,
		processVisibleIndicesEvent
	]);
	const elementObserver = useMemo(() => {
		if (typeof ResizeObserver !== "function") return;
		return new ResizeObserver(() => {
			flushSync(updateElements);
		});
	}, [updateElements]);
	useEffect(() => {
		for (const element of elements.values()) elementObserver?.observe(element);
		return () => elementObserver?.disconnect();
	}, [elementObserver, elements]);
	const itemRef = useCallback((element) => {
		if (!element) return;
		if (itemSize) return;
		const prevElement = elements.get(element.id);
		if (prevElement && prevElement !== element) elementObserver?.unobserve(prevElement);
		updateElements();
		elements.set(element.id, element);
		elementObserver?.observe(element);
	}, [
		itemSize,
		elements,
		updateElements,
		elementObserver
	]);
	const getItemProps = useCallback((item, index) => {
		const itemId = getItemId(item, index, baseId);
		const offset = itemSize ? paddingStart + itemSize * index + gap * index : data.get(itemId)?.start ?? 0;
		const baseItemProps = {
			id: itemId,
			ref: itemRef,
			index,
			style: {
				position: "absolute",
				left: horizontal ? offset : 0,
				top: horizontal ? 0 : offset
			}
		};
		if (itemSize) baseItemProps.style[horizontal ? "width" : "height"] = itemSize;
		if (item == null) return baseItemProps;
		const itemProps = getItemObject(item);
		return {
			...itemProps,
			...baseItemProps,
			style: {
				...itemProps.style,
				...baseItemProps.style
			}
		};
	}, [
		baseId,
		data,
		itemSize,
		paddingStart,
		gap,
		horizontal,
		itemRef
	]);
	const itemsProps = useMemo(() => {
		return visibleIndices.map((index) => {
			if (index < 0) return;
			const item = getItem(items, index);
			if (!item) return;
			return getItemProps(item, index);
		}).filter((value) => value != null);
	}, [
		items,
		visibleIndices,
		getItemProps
	]);
	useEffect(() => {
		const renderedIds = itemSize ? /* @__PURE__ */ new Set() : new Set(itemsProps.map((itemProps) => itemProps.id));
		for (const [id, element] of elements) {
			if (renderedIds.has(id)) continue;
			elementObserver?.unobserve(element);
			elements.delete(id);
		}
	}, [
		itemsProps,
		itemSize,
		elements,
		elementObserver
	]);
	const children = itemsProps?.map((itemProps) => {
		return renderItem?.(itemProps);
	});
	const styleProp = props.style;
	const sizeProperty = horizontal ? "width" : "height";
	const style = useMemo(() => ({
		flex: "none",
		position: "relative",
		[sizeProperty]: totalSize,
		...styleProp
	}), [
		styleProp,
		sizeProperty,
		totalSize
	]);
	const childrenData = useMemo(() => /* @__PURE__ */ new Map(), []);
	const providerValue = useMemo(() => ({
		store,
		orientation,
		overscan,
		childrenData
	}), [
		store,
		orientation,
		overscan,
		childrenData
	]);
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(CollectionRendererContext.Provider, {
		value: providerValue,
		children: element
	}), [providerValue]);
	props = {
		id: baseId,
		...props,
		style,
		ref: useMergeRefs(ref, props.ref)
	};
	return {
		...props,
		children
	};
}
const CollectionRenderer = forwardRef(function CollectionRenderer(props) {
	return createElement(TagName, useCollectionRenderer(props));
});
const getCollectionRendererItem = getItem;
const getCollectionRendererItemId = getItemId;
//#endregion
export { CollectionRenderer, getCollectionRendererItem, getCollectionRendererItemId, useCollectionRenderer };
