// @ariakit/react-components/dist/dialog/dialog-dismiss.js
"use client";
import { useButton } from "../button/button.js";
import { useDialogScopedContext } from "./dialog-context.js";
import { createElement, createHook, forwardRef, useEvent } from "@ariakit/react-utils";
import { useMemo } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
//#region src/dialog/dialog-dismiss.tsx
const TagName = "button";
/**
* Returns props to create a `DialogDismiss` component.
* @see https://ariakit.com/components/dialog
* @example
* ```jsx
* const store = useDialogStore();
* const props = useDialogDismiss({ store });
* <Dialog store={store}>
*   <Role {...props} />
* </Dialog>
* ```
*/
const useDialogDismiss = createHook(function useDialogDismiss({ store, ...props }) {
	const context = useDialogScopedContext();
	store = store || context;
	const onClickProp = props.onClick;
	const onClick = useEvent((event) => {
		onClickProp?.(event);
		if (event.defaultPrevented) return;
		store?.hide();
	});
	props = {
		"data-dialog-dismiss": "",
		children: useMemo(() => /* @__PURE__ */ jsxs("svg", {
			"aria-label": "Dismiss popup",
			display: "block",
			fill: "none",
			stroke: "currentColor",
			strokeLinecap: "round",
			strokeLinejoin: "round",
			strokeWidth: 1.5,
			viewBox: "0 0 16 16",
			height: "1em",
			width: "1em",
			children: [/* @__PURE__ */ jsx("line", {
				x1: "4",
				y1: "4",
				x2: "12",
				y2: "12"
			}), /* @__PURE__ */ jsx("line", {
				x1: "4",
				y1: "12",
				x2: "12",
				y2: "4"
			})]
		}), []),
		...props,
		onClick
	};
	props = useButton(props);
	return props;
});
/**
* Renders a button that hides a
* [`Dialog`](https://ariakit.com/reference/dialog) when clicked.
* @see https://ariakit.com/components/dialog
* @example
* ```jsx {4}
* const [open, setOpen] = useState(false);
*
* <Dialog open={open} onClose={() => setOpen(false)}>
*   <DialogDismiss />
* </Dialog>
* ```
*/
const DialogDismiss = forwardRef(function DialogDismiss(props) {
	return createElement(TagName, useDialogDismiss(props));
});
//#endregion
export { DialogDismiss, useDialogDismiss };
