// @tinacms/bridge/dist/preview.js
const PREVIEW_HEADER = "X-Tina-Preview";
const PREVIEW_CONTENT_TYPE = "application/x-tina-preview+json";
const PRIME_HEADER = "X-Tina-Prime";
const MAX_ENVELOPE_BYTES = 1e6;
async function readOverlay(request, queryId) {
  const envelope = await readEnvelope(request);
  if (!envelope)
    return void 0;
  const value = envelope[queryId];
  return value === void 0 ? void 0 : value;
}
const envelopeCache = /* @__PURE__ */ new WeakMap();
function readEnvelope(request) {
  let cached = envelopeCache.get(request);
  if (!cached) {
    cached = parseEnvelope(request);
    envelopeCache.set(request, cached);
  }
  return cached;
}
async function parseEnvelope(request) {
  const contentType = request.headers.get("content-type") ?? "";
  if (!contentType.includes(PREVIEW_CONTENT_TYPE))
    return void 0;
  const declaredLength = Number(request.headers.get("content-length") ?? "0");
  if (declaredLength > MAX_ENVELOPE_BYTES)
    return void 0;
  try {
    const text = await request.text();
    if (!text)
      return void 0;
    if (text.length > MAX_ENVELOPE_BYTES)
      return void 0;
    return JSON.parse(text);
  } catch {
    return void 0;
  }
}
export {
  PREVIEW_CONTENT_TYPE,
  PREVIEW_HEADER,
  PRIME_HEADER,
  readEnvelope,
  readOverlay
};
