// @ariakit/react-components/dist/form/form-control.js
"use client";
import { useCollectionItem } from "../collection/collection-item.js";
import { useFormItem } from "./form-context.js";
import { createElement, createHook, forwardRef, memo, useBooleanEvent, useEvent, useMergeRefs } from "@ariakit/react-utils";
import { cx, getDocument } from "@ariakit/utils";
import { useMemo } from "react";
import { useStoreStateObject } from "@ariakit/react-store";
//#region src/form/form-control.tsx
const TagName = "input";
function getNamedElement(ref, name) {
	const element = ref.current;
	if (!element) return null;
	if (element.name === name) return element;
	if (element.form) return element.form.elements.namedItem(name);
	return getDocument(element).getElementsByName(name)[0];
}
function getControlItemIds(items, name) {
	const ids = {};
	for (const item of items) {
		if (item.name !== name) continue;
		if (item.type === "label" && ids.labelId === void 0) ids.labelId = item.id;
		else if (item.type === "error" && ids.errorId === void 0) ids.errorId = item.id;
		else if (item.type === "description" && ids.descriptionId === void 0) ids.descriptionId = item.id;
		if (ids.labelId !== void 0 && ids.errorId !== void 0 && ids.descriptionId !== void 0) break;
	}
	return ids;
}
function useControlState(form, name) {
	const getItemIds = useMemo(() => {
		let prevItems;
		let prevIds = {};
		return (state) => {
			if (state.items !== prevItems) {
				prevItems = state.items;
				prevIds = getControlItemIds(state.items, name);
			}
			return prevIds;
		};
	}, [name]);
	return useStoreStateObject(form, {
		labelId: (state) => getItemIds(state).labelId,
		errorId: (state) => getItemIds(state).errorId,
		descriptionId: (state) => getItemIds(state).descriptionId,
		invalid: () => !!form.getError(name) && form.getFieldTouched(name)
	});
}
/**
* Returns props to create a `FormControl` component. Unlike `useFormInput`,
* this hook doesn't automatically returns the `value` and `onChange` props.
* This is so we can use it not only for native form elements but also for
* custom components whose value is not controlled by the native `value` and
* `onChange` props.
* @see https://ariakit.com/components/form
* @example
* ```jsx
* const store = useFormStore({ defaultValues: { content: "" } });
* const props = useFormControl({ store, name: store.names.content });
* const value = store.useValue(store.names.content);
*
* <Form store={store}>
*   <FormLabel name={store.names.content}>Content</FormLabel>
*   <Role
*     {...props}
*     value={value}
*     onChange={(value) => store.setValue(store.names.content, value)}
*     render={<Editor />}
*   />
* </Form>
* ```
*/
const useFormControl = createHook(function useFormControl({ store, name: nameProp, getItem: getItemProp, touchOnBlur = true, ...props }) {
	const { store: form, name, id, ref, getItem } = useFormItem({
		store,
		name: nameProp,
		id: props.id,
		type: "field",
		getItem: getItemProp,
		component: "FormControl"
	});
	form.useValidate(async () => {
		const element = getNamedElement(ref, name);
		if (!element) return;
		await Promise.resolve();
		if ("validity" in element && !element.validity.valid) form.setError(name, element.validationMessage);
	});
	const onBlurProp = props.onBlur;
	const touchOnBlurProp = useBooleanEvent(touchOnBlur);
	const onBlur = useEvent((event) => {
		onBlurProp?.(event);
		if (event.defaultPrevented) return;
		if (!touchOnBlurProp(event)) return;
		form.setFieldTouched(name, true);
	});
	const { labelId, errorId, descriptionId, invalid } = useControlState(form, name);
	const describedBy = cx(errorId, descriptionId, props["aria-describedby"]);
	props = {
		"aria-labelledby": props["aria-label"] != null ? void 0 : labelId,
		"aria-invalid": invalid,
		...props,
		id,
		"aria-describedby": describedBy || void 0,
		ref: useMergeRefs(ref, props.ref),
		onBlur
	};
	props = useCollectionItem({
		store: form,
		...props,
		name,
		getItem
	});
	return props;
});
/**
* Abstract component that renders a form control. Unlike
* [`FormInput`](https://ariakit.com/reference/form-input), this component
* doesn't automatically pass the `value` and `onChange` props down to the
* underlying element. This is so we can use it not only for native form
* elements but also for custom components whose value is not controlled by the
* native `value` and `onChange` props.
* @see https://ariakit.com/components/form
* @example
* ```jsx {11-19}
* const form = useFormStore({
*   defaultValues: {
*     content: "",
*   },
* });
*
* const value = form.useValue(form.names.content);
*
* <Form store={form}>
*   <FormLabel name={form.names.content}>Content</FormLabel>
*   <FormControl
*     name={form.names.content}
*     render={
*       <Editor
*         value={value}
*         onChange={(value) => form.setValue(form.names.content, value)}
*       />
*     }
*   />
* </Form>
* ```
*/
const FormControl = memo(forwardRef(function FormControl(props) {
	return createElement(TagName, useFormControl(props));
}));
//#endregion
export { FormControl, useFormControl };
