// @ariakit/react-components/dist/form/form-description.js
"use client";
import { useCollectionItem } from "../collection/collection-item.js";
import { useFormItem } from "./form-context.js";
import { createElement, createHook, forwardRef, memo, useMergeRefs } from "@ariakit/react-utils";
//#region src/form/form-description.tsx
const TagName = "div";
/**
* Returns props to create a `FormDescription` component.
* @see https://ariakit.com/components/form
* @example
* ```jsx
* const store = useFormStore({ defaultValues: { password: "" } });
* const props = useFormDescription({ store, name: store.names.password });
* <Form store={store}>
*   <FormLabel name={store.names.password}>Password</FormLabel>
*   <FormInput name={store.names.password} type="password" />
*   <Role {...props}>Password with at least 8 characters.</Role>
* </Form>
* ```
*/
const useFormDescription = createHook(function useFormDescription({ store, name: nameProp, getItem: getItemProp, ...props }) {
	const { store: form, id, ref, getItem } = useFormItem({
		store,
		name: nameProp,
		id: props.id,
		type: "description",
		getItem: getItemProp,
		component: "FormDescription"
	});
	props = {
		...props,
		id,
		ref: useMergeRefs(ref, props.ref)
	};
	props = useCollectionItem({
		store: form,
		...props,
		getItem
	});
	return props;
});
/**
* Renders a description element for a form field, which will automatically
* receive an `aria-describedby` attribute pointing to this element.
* @see https://ariakit.com/components/form
* @example
* ```jsx {10-12}
* const form = useFormStore({
*   defaultValues: {
*     password: "",
*   },
* });
*
* <Form store={form}>
*   <FormLabel name={form.names.password}>Password</FormLabel>
*   <FormInput name={form.names.password} type="password" />
*   <FormDescription name={form.names.password}>
*     Password with at least 8 characters.
*   </FormDescription>
* </Form>
* ```
*/
const FormDescription = memo(forwardRef(function FormDescription(props) {
	return createElement(TagName, useFormDescription(props));
}));
//#endregion
export { FormDescription, useFormDescription };
