// @ariakit/react-components/dist/form/form-context.js
"use client";
import { CollectionContextProvider, CollectionScopedContextProvider } from "../collection/collection-context.js";
import { createStoreContext, useId } from "@ariakit/react-utils";
import { invariant } from "@ariakit/utils";
import { useCallback, useRef } from "react";
//#region src/form/form-context.tsx
const ctx = createStoreContext([CollectionContextProvider], [CollectionScopedContextProvider]);
/**
* Returns the form store from the nearest form container.
* @example
* function FormInput() {
*   const store = useFormContext();
*
*   if (!store) {
*     throw new Error("FormInput must be wrapped in FormProvider");
*   }
*
*   // Use the store...
* }
*/
const useFormContext = ctx.useContext;
function useFormItemContext({ store, name: nameProp, component }) {
	const context = useFormContext();
	const form = store || context;
	invariant(form, process.env.NODE_ENV !== "production" && `${component} must be wrapped in a Form component.`);
	return {
		store: form,
		name: String(nameProp)
	};
}
function useFormItem({ id: idProp, type, getItem: getItemProp, ...options }) {
	const { store, name } = useFormItemContext(options);
	const id = useId(idProp);
	return {
		store,
		name,
		id,
		ref: useRef(null),
		getItem: useCallback((item) => {
			const nextItem = {
				...item,
				id: id || item.id,
				name,
				type
			};
			if (getItemProp) return getItemProp(nextItem);
			return nextItem;
		}, [
			id,
			name,
			type,
			getItemProp
		])
	};
}
const useFormScopedContext = ctx.useScopedContext;
const useFormProviderContext = ctx.useProviderContext;
const FormContextProvider = ctx.ContextProvider;
const FormScopedContextProvider = ctx.ScopedContextProvider;
//#endregion
export { FormContextProvider, FormScopedContextProvider, useFormContext, useFormItem, useFormItemContext, useFormProviderContext, useFormScopedContext };
