// @ariakit/react-components/dist/form/form-checkbox.js
"use client";
import { useCheckboxStore } from "../checkbox/checkbox-store.js";
import { n as useCheckbox } from "../__chunks/Bz1czXRN.js";
import { useFormItemContext } from "./form-context.js";
import { useFormControl } from "./form-control.js";
import { createElement, createHook, forwardRef, memo } from "@ariakit/react-utils";
//#region src/form/form-checkbox.tsx
const TagName = "input";
/**
* Returns props to create a `FormCheckbox` component.
* @see https://ariakit.com/components/form
* @example
* ```jsx
* const store = useFormStore({ defaultValues: { acceptTerms: false } });
* const props = useFormCheckbox({ store, name: store.names.acceptTerms });
* <Form store={store}>
*   <label>
*     <Role {...props} />
*     Accept terms
*   </label>
* </Form>
* ```
*/
const useFormCheckbox = createHook(function useFormCheckbox({ store, name: nameProp, value, checked, defaultChecked, ...props }) {
	const { store: form, name } = useFormItemContext({
		store,
		name: nameProp,
		component: "FormCheckbox"
	});
	props = useCheckbox({
		store: useCheckboxStore({
			value: form.useValue(name),
			setValue: (value) => form.setValue(name, value)
		}),
		value,
		checked,
		...props
	});
	props = useFormControl({
		store: form,
		name,
		"aria-labelledby": void 0,
		...props
	});
	return props;
});
/**
* Renders a checkbox input as a form control, representing a boolean, string,
* number, or array value.
* @see https://ariakit.com/components/form
* @example
* ```jsx {9}
* const form = useFormStore({
*   defaultValues: {
*     acceptTerms: false,
*   },
* });
*
* <Form store={form}>
*   <label>
*     <FormCheckbox name={form.names.acceptTerms} />
*     Accept terms
*   </label>
* </Form>
* ```
*/
const FormCheckbox = memo(forwardRef(function FormCheckbox(props) {
	return createElement(TagName, useFormCheckbox(props));
}));
//#endregion
export { FormCheckbox, useFormCheckbox };
