// @ariakit/react-components/dist/form/form-remove.js
"use client";
import { useButton } from "../button/button.js";
import { useFormItemContext } from "./form-context.js";
import { getArrayFieldIndex, isArrayFieldName } from "./utils.js";
import { createElement, createHook, forwardRef, useEvent } from "@ariakit/react-utils";
import { isTextField } from "@ariakit/utils";
//#region src/form/form-remove.tsx
const TagName = "button";
function findNextOrPreviousField(items, name, index) {
	const fields = items?.filter((item) => item.type === "field" && isArrayFieldName(item.name, name));
	const nextField = fields?.find((field) => getArrayFieldIndex(field.name, name) > index);
	if (nextField) return nextField;
	return fields?.reverse().find((field) => getArrayFieldIndex(field.name, name) < index);
}
function findPushButton(items, name) {
	return items?.find((item) => item.type === "button" && item.name === name);
}
/**
* Returns props to create a `FormRemove` component.
* @see https://ariakit.com/components/form
* @example
* ```jsx
* const store = useFormStore({
*   defaultValues: {
*     languages: ["JavaScript", "PHP"],
*   },
* });
* const props = useFormRemove({
*   store,
*   name: store.names.languages,
*   index: 0,
* });
* const values = useStoreState(store, "values");
*
* <Form store={store}>
*   {values.languages.map((language, i) => {
*     if (language == null) return null;
*     return <FormInput key={i} name={store.names.languages[i]} />;
*   })}
*   <Role {...props}>Remove first language</Role>
* </Form>
* ```
*/
const useFormRemove = createHook(function useFormRemove({ store, name: nameProp, index, autoFocusOnClick = true, ...props }) {
	const { store: form, name } = useFormItemContext({
		store,
		name: nameProp,
		component: "FormRemove"
	});
	const onClickProp = props.onClick;
	const onClick = useEvent((event) => {
		onClickProp?.(event);
		if (event.defaultPrevented) return;
		form.removeValue(name, index);
		if (!autoFocusOnClick) return;
		const { items } = form.getState();
		const element = findNextOrPreviousField(items, name, index)?.element;
		if (element) {
			element.focus();
			if (isTextField(element)) element.select();
		} else findPushButton(items, name)?.element?.focus();
	});
	props = {
		...props,
		onClick
	};
	props = useButton(props);
	return props;
});
/**
* Renders a button that will remove an item from an array field in the form
* when clicked.
*
* The [`name`](https://ariakit.com/reference/form-remove#name) prop must be
* provided to identify the array field. Similarly, the
* [`index`](https://ariakit.com/reference/form-remove#index) prop is required
* to pinpoint the item to remove.
*
* By default, the button will automatically move focus to the next field in the
* form when clicked, or to the previous field if there isn't a next field. This
* behavior can be disabled by setting the
* [`autoFocusOnClick`](https://ariakit.com/reference/form-remove#autofocusonclick)
* prop to `false`.
* @see https://ariakit.com/components/form
* @example
* ```jsx {15}
* const form = useFormStore({
*   defaultValues: {
*     languages: ["JavaScript", "PHP"],
*   },
* });
*
* const values = useStoreState(form, "values");
*
* <Form store={form}>
*   {values.languages.map((language, i) => {
*     if (language == null) return null;
*     return (
*       <div key={i}>
*         <FormInput name={form.names.languages[i]} />
*         <FormRemove name={form.names.languages} index={i} />
*       </div>
*     );
*   })}
* </Form>
* ```
*/
const FormRemove = forwardRef(function FormRemove(props) {
	return createElement(TagName, useFormRemove(props));
});
//#endregion
export { FormRemove, useFormRemove };
