// @ariakit/react-components/dist/form/form.js
"use client";
import { FormScopedContextProvider, useFormContext } from "./form-context.js";
import { createElement, createHook, forwardRef, useEvent, useInitialValue, useMergeRefs, useTagName, useUpdateEffect, useWrapElement } from "@ariakit/react-utils";
import { invariant, isTextField, sortBasedOnDOMPosition } from "@ariakit/utils";
import { useEffect, useRef, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
//#region src/form/form.tsx
const TagName = "form";
function isField(element, items) {
	return items.some((item) => item.type === "field" && item.element === element);
}
function getFirstInvalidField(items) {
	return sortBasedOnDOMPosition(items, (item) => item.element).find((item) => item.type === "field" && item.element?.getAttribute("aria-invalid") === "true");
}
/**
* Returns props to create a `Form` component.
* @see https://ariakit.com/components/form
* @example
* ```jsx
* const store = useFormStore();
* const props = useForm({ store, render: <form /> });
* <Role {...props} />
* ```
*/
const useForm = createHook(function useForm({ store, validateOnChange = true, validateOnBlur = true, resetOnUnmount = false, resetOnSubmit = true, autoFocusOnSubmit = true, ...props }) {
	const context = useFormContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "Form must receive a `store` prop or be wrapped in a FormProvider component.");
	const ref = useRef(null);
	const values = useStoreState(store, "values");
	const submitSucceed = useStoreState(store, "submitSucceed");
	const submitFailed = useStoreState(store, "submitFailed");
	const items = useStoreState(store, "items");
	const defaultValues = useInitialValue(values);
	useEffect(() => resetOnUnmount ? store?.reset : void 0, [resetOnUnmount, store]);
	useUpdateEffect(() => {
		if (!validateOnChange) return;
		if (values === defaultValues) return;
		store?.validate();
	}, [
		validateOnChange,
		values,
		defaultValues,
		store
	]);
	useEffect(() => {
		if (!resetOnSubmit) return;
		if (!submitSucceed) return;
		store?.reset();
	}, [
		resetOnSubmit,
		submitSucceed,
		store
	]);
	const [shouldFocusOnSubmit, setShouldFocusOnSubmit] = useState(false);
	useEffect(() => {
		if (!submitSucceed) return;
		setShouldFocusOnSubmit(false);
	}, [submitSucceed]);
	useEffect(() => {
		if (!shouldFocusOnSubmit) return;
		if (!submitFailed) return;
		const element = getFirstInvalidField(items)?.element;
		if (!element) return;
		setShouldFocusOnSubmit(false);
		element.focus();
		if (isTextField(element)) element.select();
	}, [
		autoFocusOnSubmit,
		submitFailed,
		items
	]);
	const onSubmitProp = props.onSubmit;
	const onSubmit = useEvent((event) => {
		onSubmitProp?.(event);
		if (event.defaultPrevented) return;
		event.preventDefault();
		store?.submit();
		if (!autoFocusOnSubmit) return;
		setShouldFocusOnSubmit(true);
	});
	const onBlurProp = props.onBlur;
	const onBlur = useEvent((event) => {
		onBlurProp?.(event);
		if (event.defaultPrevented) return;
		if (!validateOnBlur) return;
		if (!store) return;
		if (!isField(event.target, store.getState().items)) return;
		store.validate();
	});
	const onResetProp = props.onReset;
	const onReset = useEvent((event) => {
		onResetProp?.(event);
		if (event.defaultPrevented) return;
		event.preventDefault();
		store?.reset();
	});
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(FormScopedContextProvider, {
		value: store,
		children: element
	}), [store]);
	props = {
		role: useTagName(ref, TagName) !== "form" ? "form" : void 0,
		noValidate: true,
		...props,
		ref: useMergeRefs(ref, props.ref),
		onSubmit,
		onBlur,
		onReset
	};
	return props;
});
/**
* Renders a form element and provides a [form
* store](https://ariakit.com/reference/use-form-store) to its controls.
*
* The form is automatically validated on change and on blur. This behavior can
* be disabled with the
* [`validateOnChange`](https://ariakit.com/reference/form#validateonchange) and
* [`validateOnBlur`](https://ariakit.com/reference/form#validateonblur) props.
*
* When the form is submitted with errors, the first invalid field is
* automatically focused thanks to the
* [`autoFocusOnSubmit`](https://ariakit.com/reference/form#autofocusonsubmit)
* prop. If it's successful, the
* [`resetOnSubmit`](https://ariakit.com/reference/form#resetonsubmit) prop
* ensures the form is reset to its initial values as defined by the
* [`defaultValues`](https://ariakit.com/reference/use-form-store#defaultvalues)
* option on the [store](https://ariakit.com/reference/use-form-store).
* @see https://ariakit.com/components/form
* @example
* ```jsx {5-8}
* const form = useFormStore({
*   defaultValues: { username: "johndoe" },
* });
*
* <Form store={form}>
*   <FormLabel name={form.names.username}>Username</FormLabel>
*   <FormInput name={form.names.username} />
* </Form>
* ```
*/
const Form = forwardRef(function Form(props) {
	return createElement(TagName, useForm(props));
});
//#endregion
export { Form, useForm };
