// @ariakit/react-components/dist/form/form-push.js
"use client";
import { useButton } from "../button/button.js";
import { useCollectionItem } from "../collection/collection-item.js";
import { useFormItemContext } from "./form-context.js";
import { getArrayFieldIndex } from "./utils.js";
import { createElement, createHook, forwardRef, useEvent } from "@ariakit/react-utils";
import { useCallback, useEffect, useState } from "react";
import { useStoreState } from "@ariakit/react-store";
//#region src/form/form-push.tsx
const TagName = "button";
function findFirstFieldByNameAndIndex(items, name, index) {
	return items?.find((item) => item.type === "field" && getArrayFieldIndex(item.name, name) === index);
}
/**
* Returns props to create a `FormPush` component.
* @see https://ariakit.com/components/form
* @example
* ```jsx
* const store = useFormStore({
*   defaultValues: {
*     languages: ["JavaScript", "PHP"],
*   },
* });
* const props = useFormPush({
*   store,
*   name: store.names.languages,
*   value: "",
* });
* const values = useStoreState(store, "values");
*
* <Form store={store}>
*   {values.languages.map((_, i) => (
*     <FormInput key={i} name={store.names.languages[i]} />
*   ))}
*   <Role {...props}>Add new language</Role>
* </Form>
* ```
*/
const useFormPush = createHook(function useFormPush({ store, value, name: nameProp, getItem: getItemProp, autoFocusOnClick = true, ...props }) {
	const { store: form, name } = useFormItemContext({
		store,
		name: nameProp,
		component: "FormPush"
	});
	const items = useStoreState(form, "items");
	const [focusIndex, setFocusIndex] = useState(null);
	useEffect(() => {
		if (focusIndex == null) return;
		const element = findFirstFieldByNameAndIndex(items, name, focusIndex)?.element;
		if (!element) return;
		element.focus();
		setFocusIndex(null);
	}, [
		items,
		focusIndex,
		name
	]);
	const getItem = useCallback((item) => {
		const nextItem = {
			...item,
			type: "button",
			name
		};
		if (getItemProp) return getItemProp(nextItem);
		return nextItem;
	}, [name, getItemProp]);
	const onClickProp = props.onClick;
	const onClick = useEvent((event) => {
		onClickProp?.(event);
		if (event.defaultPrevented) return;
		const length = form.getValue(name)?.length ?? 0;
		form.pushValue(name, value);
		if (!autoFocusOnClick) return;
		setFocusIndex(length);
	});
	props = {
		...props,
		onClick
	};
	props = useButton(props);
	props = useCollectionItem({
		store: form,
		...props,
		getItem
	});
	return props;
});
/**
* Renders a button that will push items to an array value in the form store
* when clicked.
*
* The [`name`](https://ariakit.com/reference/form-push#name) prop needs to be
* provided to identify the array field. The
* [`value`](https://ariakit.com/reference/form-push#value) prop is required to
* define the value that will be added to the array.
*
* By default, the newly added input will be automatically focused when the
* button is clicked unless the
* [`autoFocusOnClick`](https://ariakit.com/reference/form-push#autofocusonclick)
* prop is set to `false`.
* @see https://ariakit.com/components/form
* @example
* ```jsx {13-15}
* const form = useFormStore({
*   defaultValues: {
*     languages: ["JavaScript", "PHP"],
*   },
* });
*
* const values = useStoreState(form, "values");
*
* <Form store={form}>
*   {values.languages.map((_, i) => (
*     <FormInput key={i} name={form.names.languages[i]} />
*   ))}
*   <FormPush name={form.names.languages} value="">
*     Add new language
*   </FormPush>
* </Form>
* ```
*/
const FormPush = forwardRef(function FormPush(props) {
	return createElement(TagName, useFormPush(props));
});
//#endregion
export { FormPush, useFormPush };
