// @ariakit/react-components/dist/select/select-renderer.js
"use client";
import { getCollectionRendererItem, getCollectionRendererItemId } from "../collection/collection-renderer.js";
import { useCompositeRenderer } from "../composite/composite-renderer.js";
import { useSelectContext } from "./select-context.js";
import { createElement, forwardRef } from "@ariakit/react-utils";
import { toArray } from "@ariakit/utils";
import { useMemo } from "react";
import { useStoreState } from "@ariakit/react-store";
//#region src/select/select-renderer.tsx
const TagName = "div";
function getItemObject(item) {
	if (!item || typeof item !== "object") return { value: `${item}` };
	return item;
}
function findIndicesByValue(items, value) {
	const values = toArray(value);
	const indices = [];
	for (const [index, item] of items.entries()) {
		if (indices.length === values.length) break;
		const object = getItemObject(item);
		if (object.value != null && values.includes(object.value)) indices.push(index);
		else if (object.items?.length) {
			if (findIndicesByValue(object.items, value).length) indices.push(index);
		}
	}
	return indices;
}
function useSelectRenderer({ store, persistentIndices: persistentIndicesProp, items: itemsProp, value: valueProp, ...props }) {
	const context = useSelectContext();
	store = store || context;
	const items = useStoreState(store, (state) => {
		if (!state) return itemsProp;
		if (!state.mounted) return 0;
		return itemsProp ?? state.items;
	});
	const value = useStoreState(store, (state) => valueProp ?? state?.value);
	const valueIndices = useMemo(() => {
		if (!items) return [];
		if (value == null) return [];
		if (typeof items === "number") return [];
		if (!items.length) return [];
		return findIndicesByValue(items, value);
	}, [items, value]);
	const persistentIndices = useMemo(() => {
		if (persistentIndicesProp) return [...persistentIndicesProp, ...valueIndices];
		return valueIndices;
	}, [valueIndices, persistentIndicesProp]);
	return useCompositeRenderer({
		store,
		items,
		persistentIndices,
		...props
	});
}
const SelectRenderer = forwardRef(function SelectRenderer(props) {
	return createElement(TagName, useSelectRenderer(props));
});
//#endregion
export { SelectRenderer, getCollectionRendererItem as getSelectRendererItem, getCollectionRendererItemId as getSelectRendererItemId };
