// @ariakit/react-components/dist/select/select-item.js
"use client";
import { useCompositeItem } from "../composite/composite-item.js";
import { useCompositeHover } from "../composite/composite-hover.js";
import { SelectItemCheckedContext, useSelectScopedContext } from "./select-context.js";
import { createElement, createHook, forwardRef, memo, useBooleanEvent, useEvent, useId, useWrapElement } from "@ariakit/react-utils";
import { disabledFromProps, getPopupItemRole, invariant, isDownloading, isOpeningInNewTab } from "@ariakit/utils";
import { useCallback } from "react";
import { jsx } from "react/jsx-runtime";
import { useStoreState, useStoreStateObject } from "@ariakit/react-store";
//#region src/select/select-item.tsx
const TagName = "div";
function isSelected(storeValue, itemValue) {
	if (itemValue == null) return;
	if (storeValue == null) return false;
	if (Array.isArray(storeValue)) return storeValue.includes(itemValue);
	return storeValue === itemValue;
}
/**
* Returns props to create a `SelectItem` component.
* @see https://ariakit.com/components/select
* @example
* ```jsx
* const store = useSelectStore();
* const props = useSelectItem({ store, value: "Apple" });
* <Role {...props} />
* ```
*/
const useSelectItem = createHook(function useSelectItem({ store, value, getItem: getItemProp, hideOnClick, setValueOnClick = value != null, preventScrollOnKeyDown = true, focusOnHover = true, ...props }) {
	const context = useSelectScopedContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "SelectItem must be wrapped in a SelectList or SelectPopover component.");
	const id = useId(props.id);
	const disabled = disabledFromProps(props);
	const { listElement, multiSelectable, selected, autoFocus } = useStoreStateObject(store, {
		listElement: "listElement",
		multiSelectable(state) {
			return Array.isArray(state.value);
		},
		selected(state) {
			return isSelected(state.value, value);
		},
		autoFocus(state) {
			if (value == null) return false;
			if (state.value == null) return false;
			if (state.activeId !== id && store?.item(state.activeId)) return false;
			if (Array.isArray(state.value)) return state.value[state.value.length - 1] === value;
			return state.value === value;
		}
	});
	const virtualFocus = useStoreState(store?.combobox, "virtualFocus");
	const getItem = useCallback((item) => {
		const nextItem = {
			...item,
			value: disabled ? void 0 : value
		};
		if (getItemProp) return getItemProp(nextItem);
		return nextItem;
	}, [
		disabled,
		value,
		getItemProp
	]);
	hideOnClick = hideOnClick ?? (value != null && !multiSelectable);
	const onClickProp = props.onClick;
	const setValueOnClickProp = useBooleanEvent(setValueOnClick);
	const hideOnClickProp = useBooleanEvent(hideOnClick);
	const onClick = useEvent((event) => {
		onClickProp?.(event);
		if (event.defaultPrevented) return;
		if (isDownloading(event)) return;
		if (isOpeningInNewTab(event)) return;
		if (setValueOnClickProp(event) && value != null) store?.setValue((prevValue) => {
			if (!Array.isArray(prevValue)) return value;
			if (prevValue.includes(value)) return prevValue.filter((v) => v !== value);
			return [...prevValue, value];
		});
		if (hideOnClickProp(event)) store?.hide();
	});
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(SelectItemCheckedContext.Provider, {
		value: selected ?? false,
		children: element
	}), [selected]);
	const shouldAutoFocus = props.autoFocus ?? autoFocus;
	props = {
		role: getPopupItemRole(listElement),
		"aria-selected": selected,
		children: value,
		...props,
		id,
		autoFocus: virtualFocus === false ? false : shouldAutoFocus,
		"data-autofocus": shouldAutoFocus || void 0,
		onClick
	};
	props = useCompositeItem({
		store,
		getItem,
		preventScrollOnKeyDown,
		...props
	});
	const focusOnHoverProp = useBooleanEvent(focusOnHover);
	props = useCompositeHover({
		store,
		...props,
		focusOnHover(event) {
			if (!focusOnHoverProp(event)) return false;
			return store.getState().open;
		}
	});
	return props;
});
/**
* Renders a select item inside a
* [`SelectList`](https://ariakit.com/reference/select-list) or
* [`SelectPopover`](https://ariakit.com/reference/select-popover).
*
* The `role` attribute will be automatically set on the item based on the
* list's own `role` attribute. For example, if the
* [`SelectPopover`](https://ariakit.com/reference/select-popover) component's
* `role` attribute is set to `listbox` (which is the default), the item `role`
* will be set to `option`.
*
* By default, the [`value`](https://ariakit.com/reference/select-item#value)
* prop will be rendered as the children, but this can be overriden if a custom
* children is provided.
* @see https://ariakit.com/components/select
* @example
* ```jsx {4-5}
* <SelectProvider>
*   <Select />
*   <SelectPopover>
*     <SelectItem value="Apple" />
*     <SelectItem value="Orange" />
*   </SelectPopover>
* </SelectProvider>
* ```
*/
const SelectItem = memo(forwardRef(function SelectItem(props) {
	return createElement(TagName, useSelectItem(props));
}));
//#endregion
export { SelectItem, useSelectItem };
