// @ariakit/react-components/dist/select/select-list.js
"use client";
import { isHidden } from "../disclosure/disclosure-content.js";
import { useComposite } from "../composite/composite.js";
import { useCompositeTypeahead } from "../composite/composite-typeahead.js";
import { SelectHeadingContext, SelectScopedContextProvider, useSelectContext } from "./select-context.js";
import { createElement, createHook, forwardRef, useBooleanEvent, useEvent, useId, useMergeRefs, useTransactionState, useWrapElement } from "@ariakit/react-utils";
import { invariant, isSelfTarget } from "@ariakit/utils";
import { createContext, useContext, useEffect, useMemo, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
//#region src/select/select-list.tsx
const TagName = "div";
const SelectListContext = createContext(null);
/**
* Returns props to create a `SelectList` component.
* @see https://ariakit.com/components/select
* @example
* ```jsx
* const store = useSelectStore();
* const props = useSelectList({ store });
* <Role {...props}>
*   <SelectItem value="Apple" />
*   <SelectItem value="Orange" />
* </Role>
* ```
*/
const useSelectList = createHook(function useSelectList({ store, resetOnEscape = true, hideOnEnter = true, composite, alwaysVisible, ...props }) {
	const context = useSelectContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "SelectList must receive a `store` prop or be wrapped in a SelectProvider component.");
	const id = useId(props.id);
	const value = useStoreState(store, "value");
	const multiSelectable = Array.isArray(value);
	const [defaultValue, setDefaultValue] = useState(value);
	const mounted = useStoreState(store, "mounted");
	useEffect(() => {
		if (mounted) return;
		setDefaultValue(value);
	}, [mounted, value]);
	resetOnEscape = resetOnEscape && !multiSelectable;
	const onKeyDownProp = props.onKeyDown;
	const resetOnEscapeProp = useBooleanEvent(resetOnEscape);
	const hideOnEnterProp = useBooleanEvent(hideOnEnter);
	const onKeyDown = useEvent((event) => {
		onKeyDownProp?.(event);
		if (event.defaultPrevented) return;
		if (event.key === "Escape" && resetOnEscapeProp(event)) store?.setValue(defaultValue);
		if (event.key === " " || event.key === "Enter") {
			if (isSelfTarget(event) && hideOnEnterProp(event)) {
				event.preventDefault();
				store?.hide();
			}
		}
	});
	const headingContext = useContext(SelectHeadingContext);
	const headingState = useState();
	const [headingId, setHeadingId] = headingContext || headingState;
	const headingContextValue = useMemo(() => [headingId, setHeadingId], [headingId, setHeadingId]);
	const [childStore, setChildStore] = useState(null);
	const setStore = useContext(SelectListContext);
	useEffect(() => {
		if (!setStore) return;
		setStore(store);
		return () => setStore(null);
	}, [setStore, store]);
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(SelectScopedContextProvider, {
		value: store,
		children: /* @__PURE__ */ jsx(SelectListContext.Provider, {
			value: setChildStore,
			children: /* @__PURE__ */ jsx(SelectHeadingContext.Provider, {
				value: headingContextValue,
				children: element
			})
		})
	}), [store, headingContextValue]);
	const hasCombobox = !!store.combobox;
	composite = composite ?? (!hasCombobox && childStore !== store);
	const [, setElement] = useTransactionState(composite ? store.setListElement : null);
	const hidden = isHidden(mounted, props.hidden, alwaysVisible);
	const style = hidden ? {
		...props.style,
		display: "none"
	} : props.style;
	if (composite) props = {
		role: "listbox",
		"aria-multiselectable": multiSelectable || void 0,
		...props
	};
	const labelId = useStoreState(store, (state) => headingId || state.labelElement?.id);
	props = {
		"aria-labelledby": props["aria-label"] != null ? void 0 : labelId,
		hidden,
		...props,
		id,
		ref: useMergeRefs(setElement, props.ref),
		style,
		onKeyDown
	};
	props = useComposite({
		store,
		...props,
		composite
	});
	props = useCompositeTypeahead({
		store,
		typeahead: !hasCombobox,
		...props
	});
	return props;
});
/**
* Renders a wrapper for
* [`SelectItem`](https://ariakit.com/reference/select-item) elements. This
* component may be rendered within a
* [`SelectPopover`](https://ariakit.com/reference/select-popover) component if
* there are other non-item elements inside the popover.
*
* The `aria-labelledby` prop is set to the
* [`Select`](https://ariakit.com/reference/select) element's `id` by default.
* @see https://ariakit.com/components/select
* @example
* ```jsx {5-8}
* <SelectProvider>
*   <Select />
*   <SelectPopover>
*     <SelectDismiss />
*     <SelectList>
*       <SelectItem value="Apple" />
*       <SelectItem value="Orange" />
*     </SelectList>
*   </SelectPopover>
* </SelectProvider>
* ```
*/
const SelectList = forwardRef(function SelectList(props) {
	return createElement(TagName, useSelectList(props));
});
//#endregion
export { SelectList, useSelectList };
