// @ariakit/react-components/dist/select/select.js
"use client";
import { getVisuallyHiddenStyle } from "../visually-hidden/visually-hidden.js";
import { t as getBasePlacement } from "../__chunks/B3k5_R1L.js";
import { usePopoverDisclosure } from "../popover/popover-disclosure.js";
import { useCompositeTypeahead } from "../composite/composite-typeahead.js";
import { SelectScopedContextProvider, useSelectProviderContext } from "./select-context.js";
import { SelectArrow } from "./select-arrow.js";
import { createElement, createHook, forwardRef, useBooleanEvent, useEvent, useMergeRefs, useWrapElement } from "@ariakit/react-utils";
import { getPopupRole, invariant, queueBeforeEvent, toArray } from "@ariakit/utils";
import { useEffect, useMemo, useRef, useState } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
//#region src/select/select.tsx
const TagName = "button";
function getSelectedValues(select) {
	return Array.from(select.selectedOptions).map((option) => option.value);
}
function nextWithValue(store, next) {
	return () => {
		const visitedIds = /* @__PURE__ */ new Set();
		let nextId = next();
		while (nextId) {
			const nextItem = store.item(nextId);
			if (!nextItem) return;
			if (nextItem.value != null) return nextItem.id;
			if (visitedIds.has(nextId)) return;
			visitedIds.add(nextId);
			nextId = next({ activeId: nextId });
		}
	};
}
/**
* Returns props to create a `Select` component.
* @see https://ariakit.com/components/select
* @example
* ```jsx
* const store = useSelectStore();
* const props = useSelect({ store });
* <Role {...props} />
* ```
*/
const useSelect = createHook(function useSelect({ store, name, form, required, showOnKeyDown = true, moveOnKeyDown = true, toggleOnPress = true, toggleOnClick = toggleOnPress, ...props }) {
	const context = useSelectProviderContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "Select must receive a `store` prop or be wrapped in a SelectProvider component.");
	const onKeyDownProp = props.onKeyDown;
	const showOnKeyDownProp = useBooleanEvent(showOnKeyDown);
	const moveOnKeyDownProp = useBooleanEvent(moveOnKeyDown);
	const dir = getBasePlacement(useStoreState(store, "placement"));
	const value = useStoreState(store, "value");
	const multiSelectable = Array.isArray(value);
	const onKeyDown = useEvent((event) => {
		onKeyDownProp?.(event);
		if (event.defaultPrevented) return;
		if (!store) return;
		const { orientation, items, activeId } = store.getState();
		const isVertical = orientation !== "horizontal";
		const isHorizontal = orientation !== "vertical";
		const isGrid = !!items.find((item) => !item.disabled && item.value != null)?.rowId;
		const getId = {
			ArrowUp: (isGrid || isVertical) && nextWithValue(store, store.up),
			ArrowRight: (isGrid || isHorizontal) && nextWithValue(store, store.next),
			ArrowDown: (isGrid || isVertical) && nextWithValue(store, store.down),
			ArrowLeft: (isGrid || isHorizontal) && nextWithValue(store, store.previous)
		}[event.key];
		if (getId && moveOnKeyDownProp(event)) {
			event.preventDefault();
			store.move(getId());
		}
		const isTopOrBottom = dir === "top" || dir === "bottom";
		if ({
			ArrowDown: isTopOrBottom,
			ArrowUp: isTopOrBottom,
			ArrowLeft: dir === "left",
			ArrowRight: dir === "right"
		}[event.key] && showOnKeyDownProp(event)) {
			event.preventDefault();
			store.move(activeId);
			queueBeforeEvent(event.currentTarget, "keyup", store.show);
		}
	});
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(SelectScopedContextProvider, {
		value: store,
		children: element
	}), [store]);
	const [autofill, setAutofill] = useState(false);
	const nativeSelectChangedRef = useRef(false);
	useEffect(() => {
		const nativeSelectChanged = nativeSelectChangedRef.current;
		nativeSelectChangedRef.current = false;
		if (nativeSelectChanged) return;
		setAutofill(false);
	}, [value]);
	const labelId = useStoreState(store, (state) => state.labelElement?.id);
	const label = props["aria-label"];
	const labelledBy = props["aria-labelledby"] || labelId;
	const items = useStoreState(store, (state) => {
		if (!name) return;
		return state.items;
	});
	const values = useMemo(() => {
		const itemValues = items?.flatMap((item) => item.value ?? []);
		return [...new Set(itemValues)];
	}, [items]);
	props = useWrapElement(props, (element) => {
		if (!name) return element;
		return /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsxs("select", {
			style: getVisuallyHiddenStyle(),
			tabIndex: -1,
			"aria-hidden": true,
			"aria-label": label,
			"aria-labelledby": label != null ? void 0 : labelledBy,
			name,
			form,
			required,
			disabled: props.disabled,
			value,
			multiple: multiSelectable,
			onFocus: () => store?.getState().selectElement?.focus(),
			onChange: (event) => {
				nativeSelectChangedRef.current = true;
				setAutofill(true);
				store?.setValue(multiSelectable ? getSelectedValues(event.target) : event.target.value);
			},
			children: [toArray(value).map((value) => {
				if (value == null) return null;
				if (values.includes(value)) return null;
				return /* @__PURE__ */ jsx("option", {
					value,
					children: value
				}, value);
			}), values.map((value) => /* @__PURE__ */ jsx("option", {
				value,
				children: value
			}, value))]
		}), element] });
	}, [
		store,
		label,
		labelledBy,
		name,
		form,
		required,
		value,
		multiSelectable,
		values,
		props.disabled
	]);
	const children = /* @__PURE__ */ jsxs(Fragment$1, { children: [value, /* @__PURE__ */ jsx(SelectArrow, {})] });
	const contentElement = useStoreState(store, "contentElement");
	props = {
		role: "combobox",
		"aria-autocomplete": "none",
		"aria-labelledby": props["aria-label"] != null ? void 0 : labelId,
		"aria-haspopup": getPopupRole(contentElement, "listbox"),
		"data-autofill": autofill || void 0,
		"data-name": name,
		children,
		...props,
		ref: useMergeRefs(store.setSelectElement, props.ref),
		onKeyDown
	};
	props = usePopoverDisclosure({
		store,
		toggleOnClick,
		...props
	});
	props = useCompositeTypeahead({
		store,
		...props
	});
	return props;
});
/**
* Renders a custom select element that controls the visibility of either a
* [`SelectList`](https://ariakit.com/reference/select-list) or a
* [`SelectPopover`](https://ariakit.com/reference/select-popover) component.
*
* By default, the
* [`value`](https://ariakit.com/reference/select-provider#value) state is
* rendered as the children, followed by a
* [`SelectArrow`](https://ariakit.com/reference/select-arrow) component. This
* can be customized by passing different children to the component.
* @see https://ariakit.com/components/select
* @example
* ```jsx {2}
* <SelectProvider>
*   <Select />
*   <SelectPopover>
*     <SelectItem value="Apple" />
*     <SelectItem value="Orange" />
*   </SelectPopover>
* </SelectProvider>
* ```
*/
const Select = forwardRef(function Select(props) {
	return createElement(TagName, useSelect(props));
});
//#endregion
export { Select, useSelect };
