// codemirror-graphql/variables/hint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const codemirror_1 = __importDefault(require("codemirror"));
const graphql_1 = require("graphql");
const forEachState_1 = __importDefault(require("../utils/forEachState"));
const hintList_1 = __importDefault(require("../utils/hintList"));
codemirror_1.default.registerHelper('hint', 'graphql-variables', (editor, options) => {
    const cur = editor.getCursor();
    const token = editor.getTokenAt(cur);
    const results = getVariablesHint(cur, token, options);
    if ((results === null || results === void 0 ? void 0 : results.list) && results.list.length > 0) {
        results.from = codemirror_1.default.Pos(results.from.line, results.from.ch);
        results.to = codemirror_1.default.Pos(results.to.line, results.to.ch);
        codemirror_1.default.signal(editor, 'hasCompletion', editor, results, token);
    }
    return results;
});
function getVariablesHint(cur, token, options) {
    const state = token.state.kind === 'Invalid' ? token.state.prevState : token.state;
    const { kind, step } = state;
    if (kind === 'Document' && step === 0) {
        return (0, hintList_1.default)(cur, token, [{ text: '{' }]);
    }
    const { variableToType } = options;
    if (!variableToType) {
        return;
    }
    const typeInfo = getTypeInfo(variableToType, token.state);
    if (kind === 'Document' || (kind === 'Variable' && step === 0)) {
        const variableNames = Object.keys(variableToType);
        return (0, hintList_1.default)(cur, token, variableNames.map(name => ({
            text: `"${name}": `,
            type: variableToType[name],
        })));
    }
    if ((kind === 'ObjectValue' || (kind === 'ObjectField' && step === 0)) &&
        typeInfo.fields) {
        const inputFields = Object.keys(typeInfo.fields).map(fieldName => typeInfo.fields[fieldName]);
        return (0, hintList_1.default)(cur, token, inputFields.map(field => ({
            text: `"${field.name}": `,
            type: field.type,
            description: field.description,
        })));
    }
    if (kind === 'StringValue' ||
        kind === 'NumberValue' ||
        kind === 'BooleanValue' ||
        kind === 'NullValue' ||
        (kind === 'ListValue' && step === 1) ||
        (kind === 'ObjectField' && step === 2) ||
        (kind === 'Variable' && step === 2)) {
        const namedInputType = typeInfo.type
            ? (0, graphql_1.getNamedType)(typeInfo.type)
            : undefined;
        if (namedInputType instanceof graphql_1.GraphQLInputObjectType) {
            return (0, hintList_1.default)(cur, token, [{ text: '{' }]);
        }
        if (namedInputType instanceof graphql_1.GraphQLEnumType) {
            const values = namedInputType.getValues();
            return (0, hintList_1.default)(cur, token, values.map(value => ({
                text: `"${value.name}"`,
                type: namedInputType,
                description: value.description,
            })));
        }
        if (namedInputType === graphql_1.GraphQLBoolean) {
            return (0, hintList_1.default)(cur, token, [
                { text: 'true', type: graphql_1.GraphQLBoolean, description: 'Not false.' },
                { text: 'false', type: graphql_1.GraphQLBoolean, description: 'Not true.' },
            ]);
        }
    }
}
function getTypeInfo(variableToType, tokenState) {
    const info = {
        type: null,
        fields: null,
    };
    (0, forEachState_1.default)(tokenState, state => {
        switch (state.kind) {
            case 'Variable': {
                info.type = variableToType[state.name];
                break;
            }
            case 'ListValue': {
                const nullableType = info.type ? (0, graphql_1.getNullableType)(info.type) : undefined;
                info.type =
                    nullableType instanceof graphql_1.GraphQLList ? nullableType.ofType : null;
                break;
            }
            case 'ObjectValue': {
                const objectType = info.type ? (0, graphql_1.getNamedType)(info.type) : undefined;
                info.fields =
                    objectType instanceof graphql_1.GraphQLInputObjectType
                        ? objectType.getFields()
                        : null;
                break;
            }
            case 'ObjectField': {
                const objectField = state.name && info.fields ? info.fields[state.name] : null;
                info.type = objectField === null || objectField === void 0 ? void 0 : objectField.type;
                break;
            }
        }
    });
    return info;
}
