// @ariakit/react-components/dist/combobox/combobox-disclosure.js
"use client";
import { useComboboxProviderContext } from "./combobox-context.js";
import { useDialogDisclosure } from "../dialog/dialog-disclosure.js";
import { createElement, createHook, forwardRef, useEvent, useSafeLayoutEffect } from "@ariakit/react-utils";
import { invariant } from "@ariakit/utils";
import { jsx } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
//#region src/combobox/combobox-disclosure.tsx
const TagName = "button";
const children = /* @__PURE__ */ jsx("svg", {
	"aria-hidden": "true",
	display: "block",
	fill: "none",
	stroke: "currentColor",
	strokeLinecap: "round",
	strokeLinejoin: "round",
	strokeWidth: 1.5,
	viewBox: "0 0 16 16",
	height: "1em",
	width: "1em",
	pointerEvents: "none",
	children: /* @__PURE__ */ jsx("polyline", { points: "4,6 8,10 12,6" })
});
/**
* Returns props to create a `ComboboxDisclosure` component that toggles the
* combobox popover visibility when clicked.
* @see https://ariakit.com/components/combobox
* @example
* ```jsx
* const store = useComboboxStore();
* const props = useComboboxDisclosure({ store });
* <Combobox store={store} />
* <Role {...props} />
* <ComboboxPopover store={store}>
*   <ComboboxItem value="Item 1" />
*   <ComboboxItem value="Item 2" />
*   <ComboboxItem value="Item 3" />
* </ComboboxPopover>
* ```
*/
const useComboboxDisclosure = createHook(function useComboboxDisclosure({ store, ...props }) {
	const context = useComboboxProviderContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "ComboboxDisclosure must receive a `store` prop or be wrapped in a ComboboxProvider component.");
	const onMouseDownProp = props.onMouseDown;
	const onMouseDown = useEvent((event) => {
		onMouseDownProp?.(event);
		if (event.defaultPrevented) return;
		event.preventDefault();
		store?.move(null);
	});
	const onClickProp = props.onClick;
	const onClick = useEvent((event) => {
		onClickProp?.(event);
		if (event.defaultPrevented) return;
		if (!store) return;
		const { baseElement } = store.getState();
		store.setDisclosureElement(baseElement);
	});
	const baseElement = useStoreState(store, "baseElement");
	const open = useStoreState(store, "open");
	useSafeLayoutEffect(() => {
		if (!baseElement) return;
		store.setDisclosureElement(baseElement);
	}, [store, baseElement]);
	props = {
		children,
		tabIndex: -1,
		"aria-label": open ? "Hide popup" : "Show popup",
		"aria-expanded": open,
		...props,
		onMouseDown,
		onClick
	};
	props = useDialogDisclosure({
		store,
		...props
	});
	return props;
});
/**
* Renders a combobox disclosure button that toggles the
* [`ComboboxPopover`](https://ariakit.com/reference/combobox-popover) element's
* visibility when clicked.
*
* Although this button is not tabbable, it remains accessible to screen reader
* users. On clicking, it automatically shifts focus to the
* [`Combobox`](https://ariakit.com/reference/combobox) element.
* @see https://ariakit.com/components/combobox
* @example
* ```jsx {3}
* <ComboboxProvider>
*   <Combobox />
*   <ComboboxDisclosure />
*   <ComboboxPopover>
*     <ComboboxItem value="Apple" />
*     <ComboboxItem value="Banana" />
*     <ComboboxItem value="Orange" />
*   </ComboboxPopover>
* </ComboboxProvider>
* ```
*/
const ComboboxDisclosure = forwardRef(function ComboboxDisclosure(props) {
	return createElement(TagName, useComboboxDisclosure(props));
});
//#endregion
export { ComboboxDisclosure, useComboboxDisclosure };
