// @ariakit/react-components/dist/combobox/combobox-cancel.js
"use client";
import { useButton } from "../button/button.js";
import { useComboboxProviderContext } from "./combobox-context.js";
import { createElement, createHook, forwardRef, useEvent, useWrapElement } from "@ariakit/react-utils";
import { invariant } from "@ariakit/utils";
import { Fragment } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
//#region src/combobox/combobox-cancel.tsx
const TagName = "button";
const children = /* @__PURE__ */ jsxs("svg", {
	"aria-hidden": "true",
	display: "block",
	viewBox: "0 0 16 16",
	fill: "none",
	stroke: "currentColor",
	strokeLinecap: "round",
	strokeLinejoin: "round",
	strokeWidth: 1.5,
	width: "1em",
	height: "1em",
	pointerEvents: "none",
	children: [/* @__PURE__ */ jsx("line", {
		x1: "5",
		y1: "5",
		x2: "11",
		y2: "11"
	}), /* @__PURE__ */ jsx("line", {
		x1: "5",
		y1: "11",
		x2: "11",
		y2: "5"
	})]
});
/**
* Returns props to create a `ComboboxCancel` component that clears the combobox
* input when clicked.
* @see https://ariakit.com/components/combobox
* @example
* ```jsx
* const store = useComboboxStore();
* const props = useComboboxCancel({ store });
* <Combobox store={store} />
* <Role {...props} />
* ```
*/
const useComboboxCancel = createHook(function useComboboxCancel({ store, hideWhenEmpty, ...props }) {
	const context = useComboboxProviderContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "ComboboxCancel must receive a `store` prop or be wrapped in a ComboboxProvider component.");
	const onClickProp = props.onClick;
	const onClick = useEvent((event) => {
		onClickProp?.(event);
		if (event.defaultPrevented) return;
		store?.setValue("");
		store?.move(null);
	});
	const comboboxId = useStoreState(store, (state) => state.baseElement?.id);
	const empty = useStoreState(store, (state) => state.value === "");
	props = useWrapElement(props, (element) => {
		if (!hideWhenEmpty) return element;
		if (empty) return /* @__PURE__ */ jsx(Fragment, {});
		return element;
	}, [hideWhenEmpty, empty]);
	props = {
		children,
		"aria-label": "Clear input",
		"aria-controls": comboboxId,
		...props,
		onClick
	};
	props = useButton(props);
	return props;
});
/**
* Renders a combobox cancel button that clears the combobox input value when
* clicked.
* @see https://ariakit.com/components/combobox
* @example
* ```jsx {3}
* <ComboboxProvider>
*   <Combobox />
*   <ComboboxCancel />
*   <ComboboxPopover>
*     <ComboboxItem value="Apple" />
*     <ComboboxItem value="Banana" />
*     <ComboboxItem value="Orange" />
*   </ComboboxPopover>
* </ComboboxProvider>
* ```
*/
const ComboboxCancel = forwardRef(function ComboboxCancel(props) {
	return createElement(TagName, useComboboxCancel(props));
});
//#endregion
export { ComboboxCancel, useComboboxCancel };
