// tinacms/dist/react.js
import { hashFromQuery, addMetadata } from "@tinacms/bridge/metadata";
import { addMetadata as addMetadata2, hashFromQuery as hashFromQuery2 } from "@tinacms/bridge/metadata";
import { QUICK_EDIT_CSS } from "@tinacms/bridge/quick-edit-css";
import React from "react";
import { tinaField } from "@tinacms/bridge/tina-field";
const TinaAdminOriginContext = React.createContext(
  null
);
function useTrustedAdminOrigins() {
  const configured = React.useContext(TinaAdminOriginContext);
  return React.useMemo(() => {
    if (configured == null) {
      return typeof window !== "undefined" ? [window.location.origin] : [];
    }
    return Array.isArray(configured) ? [...configured] : [configured];
  }, [configured]);
}
function isFromAdmin(event, trustedOrigins) {
  if (typeof window === "undefined")
    return false;
  if (!trustedOrigins.includes(event.origin))
    return false;
  return event.source === window.parent;
}
function useTina(props) {
  const stringifiedQuery = JSON.stringify({
    query: props.query,
    variables: props.variables
  });
  const id = React.useMemo(
    () => hashFromQuery(stringifiedQuery),
    [stringifiedQuery]
  );
  const processedData = React.useMemo(() => {
    if (props.data) {
      const dataCopy = JSON.parse(JSON.stringify(props.data));
      return addMetadata(id, dataCopy, []);
    }
  }, [props.data, id]);
  const trustedAdminOrigins = useTrustedAdminOrigins();
  const [data, setData] = React.useState(processedData);
  const [isClient, setIsClient] = React.useState(false);
  const [quickEditEnabled, setQuickEditEnabled] = React.useState(false);
  const [isInTinaIframe, setIsInTinaIframe] = React.useState(false);
  React.useEffect(() => {
    setIsClient(true);
    setData(processedData);
    parent.postMessage({
      type: "url-changed"
    });
  }, [id, processedData]);
  React.useEffect(() => {
    if (quickEditEnabled) {
      let mouseDownHandler = function(e) {
        const attributeNames = e.target.getAttributeNames();
        const tinaAttribute = attributeNames.find(
          (name) => name.startsWith("data-tina-field")
        );
        let fieldName;
        if (tinaAttribute) {
          e.preventDefault();
          e.stopPropagation();
          fieldName = e.target.getAttribute(tinaAttribute);
        } else {
          const ancestor = e.target.closest(
            "[data-tina-field], [data-tina-field-overlay]"
          );
          if (ancestor) {
            const attributeNames2 = ancestor.getAttributeNames();
            const tinaAttribute2 = attributeNames2.find(
              (name) => name.startsWith("data-tina-field")
            );
            if (tinaAttribute2) {
              e.preventDefault();
              e.stopPropagation();
              fieldName = ancestor.getAttribute(tinaAttribute2);
            }
          }
        }
        if (fieldName && isInTinaIframe) {
          parent.postMessage(
            { type: "field:selected", fieldName },
            window.location.origin
          );
        }
      };
      const style = document.createElement("style");
      style.type = "text/css";
      style.textContent = QUICK_EDIT_CSS;
      document.head.appendChild(style);
      document.body.classList.add("__tina-quick-editing-enabled");
      document.addEventListener("click", mouseDownHandler, true);
      return () => {
        document.removeEventListener("click", mouseDownHandler, true);
        document.body.classList.remove("__tina-quick-editing-enabled");
        style.remove();
      };
    }
  }, [quickEditEnabled, isInTinaIframe]);
  React.useEffect(() => {
    if (props == null ? void 0 : props.experimental___selectFormByFormId) {
      parent.postMessage({
        type: "user-select-form",
        formId: props.experimental___selectFormByFormId()
      });
    }
  }, [id]);
  React.useEffect(() => {
    const { experimental___selectFormByFormId, ...rest } = props;
    parent.postMessage({ type: "open", ...rest, id }, window.location.origin);
    const handleMessage = (event) => {
      if (!isFromAdmin(event, trustedAdminOrigins))
        return;
      if (event.data.type === "quickEditEnabled") {
        setQuickEditEnabled(event.data.value);
      }
      if (event.data.id === id && event.data.type === "updateData") {
        const rawData = event.data.data;
        const newlyProcessedData = addMetadata(
          id,
          JSON.parse(JSON.stringify(rawData)),
          []
        );
        setData(newlyProcessedData);
        setIsInTinaIframe(true);
        const anyTinaField = document.querySelector("[data-tina-field]");
        if (anyTinaField) {
          parent.postMessage(
            { type: "quick-edit", value: true },
            window.location.origin
          );
        } else {
          parent.postMessage(
            { type: "quick-edit", value: false },
            window.location.origin
          );
        }
      }
    };
    window.addEventListener("message", handleMessage);
    return () => {
      window.removeEventListener("message", handleMessage);
      parent.postMessage({ type: "close", id }, window.location.origin);
    };
  }, [id, setQuickEditEnabled, trustedAdminOrigins]);
  return { data, isClient };
}
function useEditState() {
  const [edit, setEdit] = React.useState(false);
  const trustedAdminOrigins = useTrustedAdminOrigins();
  React.useEffect(() => {
    if (typeof window === "undefined")
      return;
    parent.postMessage({ type: "isEditMode" }, window.location.origin);
    const handleMessage = (event) => {
      var _a;
      if (!isFromAdmin(event, trustedAdminOrigins))
        return;
      if (((_a = event.data) == null ? void 0 : _a.type) === "tina:editMode") {
        setEdit(true);
      }
    };
    window.addEventListener("message", handleMessage);
    return () => {
      window.removeEventListener("message", handleMessage);
    };
  }, [trustedAdminOrigins]);
  return { edit };
}
export {
  addMetadata2 as addMetadata,
  hashFromQuery2 as hashFromQuery,
  tinaField,
  useEditState,
  useTina
};
