// @ariakit/react/cjs/__chunks/4EMCVETO.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _HEDXFQCWcjs = require('./HEDXFQCW.cjs');




var _FHQLDZSWcjs = require('./FHQLDZSW.cjs');



var _5FQZLBJ7cjs = require('./5FQZLBJ7.cjs');







var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/menubar/menubar-context.js
var menubar = _LZ6L3ECGcjs.createStoreContext.call(void 0, [_FHQLDZSWcjs.CompositeContextProvider], [_FHQLDZSWcjs.CompositeScopedContextProvider]);
var useMenubarContext = menubar.useContext;
var useMenubarScopedContext = menubar.useScopedContext;
var useMenubarProviderContext = menubar.useProviderContext;
var MenubarContextProvider = menubar.ContextProvider;
var MenubarScopedContextProvider = menubar.ScopedContextProvider;

// ../ariakit-components/dist/menubar/menubar-store.js
function createMenubarStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  return _FHQLDZSWcjs.createCompositeStore.call(void 0, {
    ...props,
    orientation: _LZ6L3ECGcjs.defaultValue.call(void 0, props.orientation, syncState == null ? void 0 : syncState.orientation, "horizontal"),
    focusLoop: _LZ6L3ECGcjs.defaultValue.call(void 0, props.focusLoop, syncState == null ? void 0 : syncState.focusLoop, true)
  });
}

// ../ariakit-react-components/dist/menubar/menubar-store.js
function useMenubarStoreProps(store, update, props) {
  return _HEDXFQCWcjs.useCompositeStoreProps.call(void 0, store, update, props);
}
function useMenubarStore(props = {}) {
  const [store, update] = _5FQZLBJ7cjs.useStore.call(void 0, createMenubarStore, props);
  return useMenubarStoreProps(store, update, props);
}

// ../ariakit-react-components/dist/menubar/menubar.js
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useMenubar = _LZ6L3ECGcjs.createHook.call(void 0, function useMenubar2({ store: storeProp, composite = true, orientation: orientationProp, virtualFocus, focusLoop, rtl, ...props }) {
  const context = useMenubarProviderContext();
  storeProp = storeProp || context;
  const store = useMenubarStore({
    store: storeProp,
    orientation: orientationProp,
    virtualFocus,
    focusLoop,
    rtl
  });
  const orientation = _5FQZLBJ7cjs.useStoreState.call(void 0, store, (state) => !composite || state.orientation === "both" ? void 0 : state.orientation);
  props = _LZ6L3ECGcjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, MenubarScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  if (composite) props = {
    role: "menubar",
    "aria-orientation": orientation,
    ...props
  };
  props = _HEDXFQCWcjs.useComposite.call(void 0, {
    store,
    composite,
    ...props
  });
  return props;
});
var Menubar = _LZ6L3ECGcjs.forwardRef.call(void 0, function Menubar2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName, useMenubar(props));
});

// ../ariakit-react-components/dist/menubar/menubar-provider.js

function MenubarProvider(props = {}) {
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, MenubarContextProvider, {
    value: useMenubarStore(props),
    children: props.children
  });
}








exports.useMenubarContext = useMenubarContext; exports.useMenubarScopedContext = useMenubarScopedContext; exports.useMenubarStore = useMenubarStore; exports.useMenubar = useMenubar; exports.Menubar = Menubar; exports.MenubarProvider = MenubarProvider;
