// @ariakit/react/cjs/__chunks/6K3TPH4G.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _4Q5TABN7cjs = require('./4Q5TABN7.cjs');



var _QH4BN4EBcjs = require('./QH4BN4EB.cjs');




var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/popover/popover-dismiss.js
var TagName = "button";
var usePopoverDismiss = _LZ6L3ECGcjs.createHook.call(void 0, function usePopoverDismiss2({ store, ...props }) {
  const context = _4Q5TABN7cjs.usePopoverScopedContext.call(void 0, );
  store = store || context;
  props = _QH4BN4EBcjs.useDialogDismiss.call(void 0, {
    store,
    ...props
  });
  return props;
});
var PopoverDismiss = _LZ6L3ECGcjs.forwardRef.call(void 0, function PopoverDismiss2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName, usePopoverDismiss(props));
});

// ../ariakit-react-components/dist/popover/popover-heading.js
var TagName2 = "h1";
var usePopoverHeading = _LZ6L3ECGcjs.createHook.call(void 0, function usePopoverHeading2(props) {
  props = _QH4BN4EBcjs.useDialogHeading.call(void 0, props);
  return props;
});
var PopoverHeading = _LZ6L3ECGcjs.forwardRef.call(void 0, function PopoverHeading2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName2, usePopoverHeading(props));
});






exports.usePopoverDismiss = usePopoverDismiss; exports.PopoverDismiss = PopoverDismiss; exports.usePopoverHeading = usePopoverHeading; exports.PopoverHeading = PopoverHeading;
