// @ariakit/react/cjs/__chunks/7EHYSRTY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _JGG73QWScjs = require('./JGG73QWS.cjs');


var _7XPGOWCKcjs = require('./7XPGOWCK.cjs');



var _6K3TPH4Gcjs = require('./6K3TPH4G.cjs');




var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/hovercard/hovercard-description.js
var TagName = "p";
var useHovercardDescription = _LZ6L3ECGcjs.createHook.call(void 0, function useHovercardDescription2(props) {
  props = _7XPGOWCKcjs.usePopoverDescription.call(void 0, props);
  return props;
});
var HovercardDescription = _LZ6L3ECGcjs.forwardRef.call(void 0, function HovercardDescription2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName, useHovercardDescription(props));
});

// ../ariakit-react-components/dist/hovercard/hovercard-dismiss.js
var TagName2 = "button";
var useHovercardDismiss = _LZ6L3ECGcjs.createHook.call(void 0, function useHovercardDismiss2({ store, ...props }) {
  const context = _JGG73QWScjs.useHovercardScopedContext.call(void 0, );
  store = store || context;
  props = _6K3TPH4Gcjs.usePopoverDismiss.call(void 0, {
    store,
    ...props
  });
  return props;
});
var HovercardDismiss = _LZ6L3ECGcjs.forwardRef.call(void 0, function HovercardDismiss2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName2, useHovercardDismiss(props));
});

// ../ariakit-react-components/dist/hovercard/hovercard-heading.js
var TagName3 = "h1";
var useHovercardHeading = _LZ6L3ECGcjs.createHook.call(void 0, function useHovercardHeading2(props) {
  props = _6K3TPH4Gcjs.usePopoverHeading.call(void 0, props);
  return props;
});
var HovercardHeading = _LZ6L3ECGcjs.forwardRef.call(void 0, function HovercardHeading2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName3, useHovercardHeading(props));
});








exports.useHovercardDescription = useHovercardDescription; exports.HovercardDescription = HovercardDescription; exports.useHovercardDismiss = useHovercardDismiss; exports.HovercardDismiss = HovercardDismiss; exports.useHovercardHeading = useHovercardHeading; exports.HovercardHeading = HovercardHeading;
