// @ariakit/react/cjs/__chunks/7YDHYTZA.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _DWSXHZYMcjs = require('./DWSXHZYM.cjs');






var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/heading/heading.js
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "h1";
var useHeading = _LZ6L3ECGcjs.createHook.call(void 0, function useHeading2(props) {
  const ref = _react.useRef.call(void 0, null);
  const level = _react.useContext.call(void 0, _DWSXHZYMcjs.HeadingContext) || 1;
  const Element = `h${level}`;
  const tagName = _LZ6L3ECGcjs.useTagName.call(void 0, ref, Element);
  const isNativeHeading = !!tagName && /^h\d$/.test(tagName);
  props = {
    render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Element, {}),
    role: !isNativeHeading ? "heading" : void 0,
    "aria-level": !isNativeHeading ? level : void 0,
    ...props,
    ref: _LZ6L3ECGcjs.useMergeRefs.call(void 0, ref, props.ref)
  };
  return props;
});
var Heading = _LZ6L3ECGcjs.forwardRef.call(void 0, function Heading2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName, useHeading(props));
});




exports.useHeading = useHeading; exports.Heading = Heading;
