// @ariakit/react/cjs/__chunks/MPRQURFL.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _4Q5TABN7cjs = require('./4Q5TABN7.cjs');



var _FHQLDZSWcjs = require('./FHQLDZSW.cjs');


var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/combobox/combobox-context.js
var _react = require('react');
var ComboboxListRoleContext = _react.createContext.call(void 0, void 0);
var ctx = _LZ6L3ECGcjs.createStoreContext.call(void 0, [_4Q5TABN7cjs.PopoverContextProvider, _FHQLDZSWcjs.CompositeContextProvider], [_4Q5TABN7cjs.PopoverScopedContextProvider, _FHQLDZSWcjs.CompositeScopedContextProvider]);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, void 0);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);










exports.ComboboxListRoleContext = ComboboxListRoleContext; exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
