// @ariakit/react/cjs/__chunks/QH4BN4EB.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _7YDHYTZAcjs = require('./7YDHYTZA.cjs');



var _ITEGEFBHcjs = require('./ITEGEFBH.cjs');


var _I4KB2YNAcjs = require('./I4KB2YNA.cjs');







var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/dialog/dialog-dismiss.js
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useDialogDismiss = _LZ6L3ECGcjs.createHook.call(void 0, function useDialogDismiss2({ store, ...props }) {
  const context = _ITEGEFBHcjs.useDialogScopedContext.call(void 0, );
  store = store || context;
  const onClickProp = props.onClick;
  const onClick = _LZ6L3ECGcjs.useEvent.call(void 0, (event) => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    store == null ? void 0 : store.hide();
  });
  props = {
    "data-dialog-dismiss": "",
    children: _react.useMemo.call(void 0, () => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, "svg", {
      "aria-label": "Dismiss popup",
      display: "block",
      fill: "none",
      stroke: "currentColor",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      strokeWidth: 1.5,
      viewBox: "0 0 16 16",
      height: "1em",
      width: "1em",
      children: [/* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", {
        x1: "4",
        y1: "4",
        x2: "12",
        y2: "12"
      }), /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", {
        x1: "4",
        y1: "12",
        x2: "12",
        y2: "4"
      })]
    }), []),
    ...props,
    onClick
  };
  props = _I4KB2YNAcjs.useButton.call(void 0, props);
  return props;
});
var DialogDismiss = _LZ6L3ECGcjs.forwardRef.call(void 0, function DialogDismiss2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName, useDialogDismiss(props));
});

// ../ariakit-react-components/dist/dialog/dialog-heading.js

var TagName2 = "h1";
var useDialogHeading = _LZ6L3ECGcjs.createHook.call(void 0, function useDialogHeading2({ store, ...props }) {
  const setHeadingId = _react.useContext.call(void 0, _ITEGEFBHcjs.DialogHeadingContext);
  const id = _LZ6L3ECGcjs.useId.call(void 0, props.id);
  _LZ6L3ECGcjs.useSafeLayoutEffect.call(void 0, () => {
    setHeadingId == null ? void 0 : setHeadingId(id);
    return () => setHeadingId == null ? void 0 : setHeadingId(void 0);
  }, [setHeadingId, id]);
  props = {
    ...props,
    id
  };
  props = _7YDHYTZAcjs.useHeading.call(void 0, props);
  return props;
});
var DialogHeading = _LZ6L3ECGcjs.forwardRef.call(void 0, function DialogHeading2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName2, useDialogHeading(props));
});






exports.useDialogDismiss = useDialogDismiss; exports.DialogDismiss = DialogDismiss; exports.useDialogHeading = useDialogHeading; exports.DialogHeading = DialogHeading;
