// @ariakit/react/cjs/__chunks/SBAK37L7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _XM764FDRcjs = require('./XM764FDR.cjs');








var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/collection/collection-item.js
var _react = require('react');
var TagName = "div";
var useCollectionItem = _LZ6L3ECGcjs.createHook.call(void 0, function useCollectionItem2({ store, shouldRegisterItem = true, getItem = _LZ6L3ECGcjs.identity, element, ...props }) {
  const context = _XM764FDRcjs.useCollectionContext.call(void 0, );
  store = store || context;
  const id = _LZ6L3ECGcjs.useId.call(void 0, props.id);
  const ref = _react.useRef.call(void 0, element);
  _react.useEffect.call(void 0, () => {
    const element2 = ref.current;
    if (!id) return;
    if (!element2) return;
    if (!shouldRegisterItem) return;
    const item = getItem({
      id,
      element: element2
    });
    return store == null ? void 0 : store.renderItem(item);
  }, [
    id,
    shouldRegisterItem,
    getItem,
    store
  ]);
  props = {
    ...props,
    ref: _LZ6L3ECGcjs.useMergeRefs.call(void 0, ref, props.ref)
  };
  return _LZ6L3ECGcjs.removeUndefinedValues.call(void 0, props);
});
var CollectionItem = _LZ6L3ECGcjs.forwardRef.call(void 0, function CollectionItem2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName, useCollectionItem(props));
});




exports.useCollectionItem = useCollectionItem; exports.CollectionItem = CollectionItem;
