// @ariakit/react/cjs/__chunks/TRMV6QE4.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _TWNFJ2SKcjs = require('./TWNFJ2SK.cjs');


var _AIFRJ53Zcjs = require('./AIFRJ53Z.cjs');




var _4Q5TABN7cjs = require('./4Q5TABN7.cjs');


var _YAZY4X3Pcjs = require('./YAZY4X3P.cjs');


var _5FQZLBJ7cjs = require('./5FQZLBJ7.cjs');








var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/popover/popover-disclosure.js
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var usePopoverDisclosure = _LZ6L3ECGcjs.createHook.call(void 0, function usePopoverDisclosure2({ store, ...props }) {
  const context = _4Q5TABN7cjs.usePopoverProviderContext.call(void 0, );
  store = store || context;
  _LZ6L3ECGcjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "PopoverDisclosure must receive a `store` prop or be wrapped in a PopoverProvider component.");
  const onClickProp = props.onClick;
  const onClick = _LZ6L3ECGcjs.useEvent.call(void 0, (event) => {
    store == null ? void 0 : store.setAnchorElement(event.currentTarget);
    onClickProp == null ? void 0 : onClickProp(event);
  });
  props = _LZ6L3ECGcjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _4Q5TABN7cjs.PopoverScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  props = {
    ...props,
    onClick
  };
  props = _TWNFJ2SKcjs.usePopoverAnchor.call(void 0, {
    store,
    ...props
  });
  props = _YAZY4X3Pcjs.useDialogDisclosure.call(void 0, {
    store,
    ...props
  });
  return props;
});
var PopoverDisclosure = _LZ6L3ECGcjs.forwardRef.call(void 0, function PopoverDisclosure2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName, usePopoverDisclosure(props));
});

// ../ariakit-react-components/dist/popover/popover-disclosure-arrow.js
var _react = require('react');

var TagName2 = "span";
var pointsMap = {
  top: "4,10 8,6 12,10",
  right: "6,4 10,8 6,12",
  bottom: "4,6 8,10 12,6",
  left: "10,4 6,8 10,12"
};
var usePopoverDisclosureArrow = _LZ6L3ECGcjs.createHook.call(void 0, function usePopoverDisclosureArrow2({ store, placement, ...props }) {
  const context = _4Q5TABN7cjs.usePopoverContext.call(void 0, );
  store = store || context;
  _LZ6L3ECGcjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "PopoverDisclosureArrow must be wrapped in a PopoverDisclosure component.");
  const points = pointsMap[_AIFRJ53Zcjs.getBasePlacement.call(void 0, _5FQZLBJ7cjs.useStoreState.call(void 0, store, (state) => placement || state.placement))];
  props = {
    children: _react.useMemo.call(void 0, () => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "svg", {
      display: "block",
      fill: "none",
      stroke: "currentColor",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      strokeWidth: 1.5,
      viewBox: "0 0 16 16",
      height: "1em",
      width: "1em",
      children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "polyline", { points })
    }), [points]),
    "aria-hidden": true,
    ...props,
    style: {
      width: "1em",
      height: "1em",
      pointerEvents: "none",
      ...props.style
    }
  };
  return _LZ6L3ECGcjs.removeUndefinedValues.call(void 0, props);
});
var PopoverDisclosureArrow = _LZ6L3ECGcjs.forwardRef.call(void 0, function PopoverDisclosureArrow2(props) {
  return _LZ6L3ECGcjs.createElement.call(void 0, TagName2, usePopoverDisclosureArrow(props));
});






exports.usePopoverDisclosure = usePopoverDisclosure; exports.PopoverDisclosure = PopoverDisclosure; exports.usePopoverDisclosureArrow = usePopoverDisclosureArrow; exports.PopoverDisclosureArrow = PopoverDisclosureArrow;
