// @ariakit/react/cjs/__chunks/WSDL6NSY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _4Q5TABN7cjs = require('./4Q5TABN7.cjs');



var _FHQLDZSWcjs = require('./FHQLDZSW.cjs');


var _LZ6L3ECGcjs = require('./LZ6L3ECG.cjs');

// ../ariakit-react-components/dist/select/select-context.js
var _react = require('react');
var ctx = _LZ6L3ECGcjs.createStoreContext.call(void 0, [_4Q5TABN7cjs.PopoverContextProvider, _FHQLDZSWcjs.CompositeContextProvider], [_4Q5TABN7cjs.PopoverScopedContextProvider, _FHQLDZSWcjs.CompositeScopedContextProvider]);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);
var SelectHeadingContext = _react.createContext.call(void 0, null);









exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext; exports.SelectHeadingContext = SelectHeadingContext;
