// @ariakit/react-components/dist/composite/composite-overflow.js
"use client";
import { usePopover } from "../popover/popover.js";
import { createElement, createHook, forwardRef, useEvent } from "@ariakit/react-utils";
import { useStoreState } from "@ariakit/react-store";
//#region src/composite/composite-overflow.tsx
const TagName = "div";
const hiddenStyle = {
	opacity: 0,
	pointerEvents: "none"
};
/**
* Returns props to create a `CompositeOverflow` component.
* @see https://ariakit.com/components/composite
* @example
* ```jsx
* const store = useCompositeOverflowStore();
* const props = useCompositeOverflow({ store });
* <Role {...props}>
*   <CompositeItem>Item 3</CompositeItem>
*   <CompositeItem>Item 4</CompositeItem>
* </Role>
* ```
*/
const useCompositeOverflow = createHook(function useCompositeOverflow({ store, wrapperProps: wrapperPropsProp, portal = false, ...props }) {
	const onFocusProp = props.onFocus;
	const onFocus = useEvent((event) => {
		onFocusProp?.(event);
		if (event.defaultPrevented) return;
		store.show();
	});
	const mounted = useStoreState(store, "mounted");
	const getStyle = (styleProp) => mounted ? styleProp : {
		...hiddenStyle,
		...styleProp
	};
	const wrapperProps = {
		...wrapperPropsProp,
		style: getStyle(wrapperPropsProp?.style)
	};
	props = {
		role: "presentation",
		...props,
		onFocus
	};
	props = usePopover({
		store,
		focusable: false,
		alwaysVisible: true,
		wrapperProps,
		portal,
		...props
	});
	return props;
});
/**
* Renders a popover that will contain the overflow items in a composite
* collection.
* @see https://ariakit.com/components/composite
* @example
* ```jsx
* const composite = useCompositeStore();
* const overflow = useCompositeOverflowStore();
* <Composite store={composite}>
*   <CompositeItem>Item 1</CompositeItem>
*   <CompositeItem>Item 2</CompositeItem>
*   <CompositeOverflowDisclosure store={overflow}>
*     +2 items
*   </CompositeOverflowDisclosure>
*   <CompositeOverflow store={overflow}>
*     <CompositeItem>Item 3</CompositeItem>
*     <CompositeItem>Item 4</CompositeItem>
*   </CompositeOverflow>
* </Composite>
* ```
*/
const CompositeOverflow = forwardRef(function CompositeOverflow(props) {
	return createElement(TagName, useCompositeOverflow(props));
});
//#endregion
export { CompositeOverflow, useCompositeOverflow };
