// @ariakit/react-components/dist/composite/composite-renderer.js
"use client";
import { getCollectionRendererItem, getCollectionRendererItemId, useCollectionRenderer } from "../collection/collection-renderer.js";
import { useCompositeScopedContext } from "./composite-context.js";
import { createElement, forwardRef, useId } from "@ariakit/react-utils";
import { useMemo } from "react";
import { useStoreState } from "@ariakit/react-store";
//#region src/composite/composite-renderer.tsx
const TagName = "div";
function getItemObject(item) {
	if (!item || typeof item !== "object") return { value: item };
	return item;
}
function countItems(items) {
	if (!items) return [0];
	if (typeof items === "number") return Array.from({ length: items }, (_, index) => index + 1);
	return items.reduce((count, item, index) => {
		const prevCount = count[index - 1] ?? 0;
		const object = getItemObject(item);
		if (!object.items) {
			count[index] = prevCount + 1;
			return count;
		}
		count[index] = prevCount + (countItems(object.items)[object.items.length - 1] ?? 0);
		return count;
	}, []);
}
function findFirst(items, offset = 1) {
	for (let index = offset > 0 ? 0 : items.length - 1; index >= 0 && index < items.length; index += offset) {
		const item = items[index];
		const object = getItemObject(item);
		if (object.items && findFirst(object.items, offset) !== -1) return index;
		if (!object.disabled) return index;
	}
	return -1;
}
function findLast(items) {
	return findFirst(items, -1);
}
function findExactById(items, id, baseId) {
	return items.findIndex((item, index) => {
		const itemId = getCollectionRendererItemId(item, index, baseId);
		if (itemId === id) return true;
		const object = getItemObject(item);
		if (object.items?.length) return findExactById(object.items, id, itemId) !== -1;
		return false;
	});
}
function findPrefixById(items, id, baseId) {
	const result = {
		index: -1,
		length: -1
	};
	items.forEach((item, index) => {
		const itemId = getCollectionRendererItemId(item, index, baseId);
		let length = id.startsWith(`${itemId}/`) ? itemId.length : -1;
		const object = getItemObject(item);
		if (object.items?.length) {
			const nestedResult = findPrefixById(object.items, id, itemId);
			length = Math.max(length, nestedResult.length);
		}
		if (length > result.length) {
			result.index = index;
			result.length = length;
		}
	});
	return result;
}
function findById(items, id, baseId) {
	const exactIndex = findExactById(items, id, baseId);
	if (exactIndex !== -1) return exactIndex;
	return findPrefixById(items, id, baseId).index;
}
function useCompositeRenderer({ store, orientation: orientationProp, persistentIndices: persistentIndicesProp, children: renderItem, "aria-setsize": ariaSetSize, "aria-posinset": ariaPosInSet = 1, ...props }) {
	const context = useCompositeScopedContext();
	store = store || context;
	const orientation = useStoreState(store, (state) => orientationProp ?? (state?.orientation === "both" ? "vertical" : state?.orientation));
	const items = useStoreState(store, (state) => {
		if (!state) return props.items;
		if ("mounted" in state && !state.mounted) return 0;
		return props.items ?? state.items;
	});
	const id = useId(props.id);
	const itemsCount = useMemo(() => countItems(items), [items]);
	const setSize = useMemo(() => ariaSetSize ?? itemsCount[itemsCount.length - 1] ?? 0, [ariaSetSize, itemsCount]);
	const firstIndex = useMemo(() => {
		if (!items) return -1;
		if (typeof items === "number") return 0;
		if (!items.length) return -1;
		return findFirst(items);
	}, [items]);
	const lastIndex = useMemo(() => {
		if (!items) return -1;
		if (typeof items === "number") return items - 1;
		if (!items.length) return -1;
		return findLast(items);
	}, [items]);
	const activeId = useStoreState(store, "activeId");
	const activeIndex = useMemo(() => {
		if (!id) return -1;
		if (!items) return -1;
		if (activeId == null) return -1;
		if (typeof items === "number") return -1;
		if (!items.length) return -1;
		return findById(items, activeId, id);
	}, [
		id,
		items,
		activeId
	]);
	const persistentIndices = useMemo(() => {
		const indices = [
			firstIndex,
			activeIndex,
			lastIndex
		].filter((index) => index >= 0);
		if (persistentIndicesProp) return [...persistentIndicesProp, ...indices];
		return indices;
	}, [
		firstIndex,
		activeIndex,
		lastIndex,
		persistentIndicesProp
	]);
	return useCollectionRenderer({
		id,
		store,
		orientation,
		persistentIndices,
		...props,
		children: (item) => {
			const nextItem = Object.assign({}, item, {
				"aria-setsize": setSize,
				"aria-posinset": ariaPosInSet + (itemsCount[item.index - 1] ?? 0)
			});
			return renderItem?.(nextItem);
		}
	});
}
const CompositeRenderer = forwardRef(function CompositeRenderer(props) {
	return createElement(TagName, useCompositeRenderer(props));
});
//#endregion
export { CompositeRenderer, getCollectionRendererItem as getCompositeRendererItem, getCollectionRendererItemId as getCompositeRendererItemId, useCompositeRenderer };
