// @ariakit/components/dist/form/form-store.js
import { createCollectionStore } from "../collection/collection-store.js";
import { createStore, init, setup, sync, throwOnConflictingProps } from "@ariakit/store";
import { applyState, defaultValue, isObject } from "@ariakit/utils";
//#region src/form/form-store.ts
const maxArrayIndex = 2 ** 32 - 2;
function getPath(path) {
	if (!Array.isArray(path)) return String(path).split(".");
	return path.map((key) => {
		if (typeof key === "number") return key;
		return String(key);
	});
}
function isPrototypePathSegment(key) {
	if (key === "__proto__") return true;
	if (key === "constructor") return true;
	return false;
}
function nextFrame() {
	return new Promise((resolve) => {
		const timeoutId = setTimeout(resolve, 100);
		requestAnimationFrame(() => {
			clearTimeout(timeoutId);
			resolve();
		});
	});
}
function hasMessages(object) {
	const keys = Object.keys(object);
	for (const key of keys) {
		if (isPrototypePathSegment(key)) continue;
		if (isObject(object[key])) {
			if (hasMessages(object[key])) return true;
			continue;
		}
		if (object[key]) return true;
	}
	return false;
}
function get(values, path, defaultValue) {
	return getPathValue(values, getPath(path), defaultValue);
}
function getPathValue(values, path, defaultValue) {
	const [key, ...rest] = path;
	if (key == null || !values || isPrototypePathSegment(key)) return defaultValue;
	if (!rest.length) return values[key] ?? defaultValue;
	return getPathValue(values[key], rest, defaultValue);
}
function isArrayIndex(key) {
	if (typeof key !== "string" && typeof key !== "number") return false;
	const stringKey = String(key);
	const index = Number(stringKey);
	return Number.isInteger(index) && index >= 0 && index <= maxArrayIndex && String(index) === stringKey;
}
function getOrCreateNested(nestedValues, nextKey) {
	if (Array.isArray(nestedValues) || isObject(nestedValues)) return nestedValues;
	return isArrayIndex(nextKey) ? [] : {};
}
function setObjectValue(values, key, value) {
	const result = {};
	if (values) {
		const keys = Object.keys(values);
		for (const propertyKey of keys) {
			if (isPrototypePathSegment(propertyKey)) continue;
			result[propertyKey] = values[propertyKey];
		}
	}
	result[key] = value;
	return result;
}
function set(values, path, value) {
	const pathKeys = getPath(path);
	if (pathKeys.some(isPrototypePathSegment)) return values;
	return setPath(values, pathKeys, value);
}
function setPath(values, path, value) {
	const [k, ...rest] = path;
	if (k == null) return values;
	const key = k;
	const isArrayIndexKey = isArrayIndex(k);
	const nestedValues = (isArrayIndexKey ? values || [] : values || {})[key];
	const nextKey = rest[0];
	const result = rest.length && nextKey != null ? setPath(getOrCreateNested(nestedValues, nextKey), rest, value) : value;
	if (isArrayIndexKey) {
		const index = Number(key);
		if (values && Array.isArray(values)) {
			const copy = [...values];
			copy[index] = result;
			return copy;
		}
		const nextValues = [];
		nextValues[index] = result;
		return nextValues;
	}
	return setObjectValue(values, key, result);
}
function setAll(values, value) {
	const result = {};
	const keys = Object.keys(values);
	for (const key of keys) {
		if (isPrototypePathSegment(key)) continue;
		const currentValue = values[key];
		if (Array.isArray(currentValue)) result[key] = currentValue.map((v) => {
			if (isObject(v)) return setAll(v, value);
			return value;
		});
		else if (isObject(currentValue)) result[key] = setAll(currentValue, value);
		else result[key] = value;
	}
	return result;
}
function getNameHandler(cache, prevKeys = []) {
	return { get(target, key) {
		if (typeof key === "symbol") {
			if (key === Symbol.toPrimitive) return () => prevKeys.join(".");
			return;
		}
		if (key === "toString") return () => prevKeys.join(".");
		if (key === "valueOf") return () => prevKeys.join(".");
		const nextKeys = [...prevKeys, key];
		const nextKey = nextKeys.join(".");
		if (cache[nextKey]) return cache[nextKey];
		const nextProxy = new Proxy(target, getNameHandler(cache, nextKeys));
		cache[nextKey] = nextProxy;
		return nextProxy;
	} };
}
function getStoreCallbacks(store) {
	return store?.__unstableCallbacks;
}
function createNames() {
	return new Proxy(Object.create(null), getNameHandler(Object.create(null)));
}
function createFormStore(props = {}) {
	throwOnConflictingProps(props, props.store);
	const syncState = props.store?.getState();
	const collection = createCollectionStore(props);
	const values = defaultValue(props.values, syncState?.values, props.defaultValues, {});
	const errors = defaultValue(props.errors, syncState?.errors, props.defaultErrors, {});
	const touched = defaultValue(props.touched, syncState?.touched, props.defaultTouched, {});
	const form = createStore({
		...collection.getState(),
		values,
		errors,
		touched,
		validating: defaultValue(syncState?.validating, false),
		submitting: defaultValue(syncState?.submitting, false),
		submitSucceed: defaultValue(syncState?.submitSucceed, 0),
		submitFailed: defaultValue(syncState?.submitFailed, 0),
		valid: !hasMessages(errors)
	}, collection, props.store);
	const syncCallbacks = getStoreCallbacks(props.store);
	const syncCallbacksState = syncCallbacks?.getState();
	const callbacks = createStore({
		validate: syncCallbacksState?.validate || [],
		submit: syncCallbacksState?.submit || []
	}, syncCallbacks);
	setup(form, () => init(callbacks));
	setup(form, () => sync(form, ["validating", "errors"], (state) => {
		if (state.validating) return;
		form.setState("valid", !hasMessages(state.errors));
	}));
	const validate = async () => {
		form.setState("validating", true);
		form.setState("errors", {});
		const validateCallbacks = callbacks.getState().validate;
		try {
			for (const callback of validateCallbacks) await callback(form.getState());
			await nextFrame();
			return !hasMessages(form.getState().errors);
		} finally {
			form.setState("validating", false);
		}
	};
	return {
		...collection,
		...form,
		names: createNames(),
		setValues: (values) => form.setState("values", values),
		getValue: (name) => get(form.getState().values, name),
		setValue: (name, value) => form.setState("values", (values) => {
			const prevValue = get(values, name);
			const nextValue = applyState(value, prevValue);
			if (nextValue === prevValue) return values;
			return set(values, name, nextValue);
		}),
		pushValue: (name, value) => form.setState("values", (values) => {
			return set(values, name, [...get(values, name, []), value]);
		}),
		removeValue: (name, index) => form.setState("values", (values) => {
			const array = get(values, name, []);
			return set(values, name, [
				...array.slice(0, index),
				null,
				...array.slice(index + 1)
			]);
		}),
		setErrors: (errors) => form.setState("errors", errors),
		getError: (name) => get(form.getState().errors, name),
		setError: (name, error) => form.setState("errors", (errors) => {
			const prevError = get(errors, name);
			const nextError = applyState(error, prevError);
			if (nextError === prevError) return errors;
			return set(errors, name, nextError);
		}),
		setTouched: (touched) => form.setState("touched", touched),
		getFieldTouched: (name) => !!get(form.getState().touched, name),
		setFieldTouched: (name, value) => form.setState("touched", (touched) => {
			const prevValue = get(touched, name);
			const nextValue = applyState(value, prevValue);
			if (nextValue === prevValue) return touched;
			return set(touched, name, nextValue);
		}),
		onValidate: (callback) => {
			callbacks.setState("validate", (callbacks) => [...callbacks, callback]);
			return () => {
				callbacks.setState("validate", (callbacks) => callbacks.filter((c) => c !== callback));
			};
		},
		validate,
		onSubmit: (callback) => {
			callbacks.setState("submit", (callbacks) => [...callbacks, callback]);
			return () => {
				callbacks.setState("submit", (callbacks) => callbacks.filter((c) => c !== callback));
			};
		},
		submit: async () => {
			form.setState("submitting", true);
			form.setState("touched", setAll(form.getState().values, true));
			try {
				if (await validate()) {
					const submitCallbacks = callbacks.getState().submit;
					for (const callback of submitCallbacks) await callback(form.getState());
					await nextFrame();
					if (!hasMessages(form.getState().errors)) {
						form.setState("submitSucceed", (count) => count + 1);
						return true;
					}
				}
				form.setState("submitFailed", (count) => count + 1);
				return false;
			} catch (error) {
				form.setState("submitFailed", (count) => count + 1);
				throw error;
			} finally {
				form.setState("submitting", false);
			}
		},
		reset: () => {
			form.setState("values", values);
			form.setState("errors", errors);
			form.setState("touched", touched);
			form.setState("validating", false);
			form.setState("submitting", false);
			form.setState("submitSucceed", 0);
			form.setState("submitFailed", 0);
			form.setState("valid", !hasMessages(errors));
		},
		__unstableCallbacks: callbacks
	};
}
//#endregion
export { createFormStore, get, hasMessages };
